
# Rcongas <a href='militeee.github.io/Rcongas'><img src='man/figures/logo.png' align="right" height="120" /></a>

<!-- badges: start -->

[![Travis build
master](https://travis-ci.org/militeee/Rcongas.svg?branch=master)](https://travis-ci.org/militeee/Rcongas)
[![Travis build
development](https://travis-ci.org/militeee/Rcongas.svg?branch=development)](https://travis-ci.org/militeee/Rcongas)
[![Lifecycle:
maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![](https://img.shields.io/badge/Part%20of-evoverse-blue.svg)](https://caravagn.github.io/evoverse)
<!-- badges: end -->



The package provides an interface for the python package ![https://github.com/Militeee/congas](Congas), a CNV caller from scRNA sequencing data.The package implements S3 objects to visualize
the data and the fits.

The associated paper can be found as:
Genotyping Copy Number Alterations from single-cell RNA sequencing. Salvatore Milite, Riccardo Bergamin, Giulio Caravagna


#### Help and support

[![](https://img.shields.io/badge/GitHub%20Pages-https://militeee.github.io/Rcongas/-steelblue.svg)](https://militeee.github.io/rcongas)

### Installation

You can install the released version of `Rcongas` from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("militeee/Rcongas")
```

-----

#### Copyright and contacts

Salvatore Milite, MSc, and Giulio Caravagna, PhD. *University of
Trieste, Trieste, Italy*.

[![](https://img.shields.io/badge/Email-gcaravagn@gmail.com-informational.svg?style=social)](mailto:gcaravagn@gmail.com)
[![](https://img.shields.io/badge/caravagn-informational.svg?style=social&logo=GitHub)](https://github.com/caravagn)
[![](https://img.shields.io/badge/@gcaravagna-informational.svg?style=social&logo=Twitter)](https://twitter.com/gcaravagna)
[![](https://img.shields.io/badge/Homepage-informational.svg?style=social&logo=Google)](https://sites.google.com/site/giuliocaravagna/)
