---
title: "Rcongas"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Rcongas}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r, warning=FALSE, message=FALSE}
library(Rcongas)
library(dplyr)
```

# Analysis workflow

These data are required:

- single-cell RNA or ATAC sequencing data from independent cells, in the form of counts per cell;
- genome-wide segmentation with associated per-segment ploidy.

The formats of the inputs is discussed here.

The (R)CONGAS+ workflow consists in:

- (optional) applying [filters](Filtering.html) on the input data;
- creating an (R)CONGAS+ object;
- (optional) applying [filters](Filtering.html) on the mapped data;
- fitting a model

The package comes with input data to show what the required input formats are, and ready to use
objects for case studies that we have discussed in the main papers; see [References](../docs/reference/index.html) for details.

# Input formats

```{r}
# Example data available in the package
data('example_input')
```

RNA and ATAC have the same input format reporting:

- `chr`, `from` and `to`: genomic locations of a gene (RNA) or a peak (ATAC)
- `cell`: a cell identifier
- `value`: the observed counts (discrete)

```{r}
# RNA
example_input$rna %>% print()
```

The `gene` column in the RNA dataset is required only to use some gene-based pre-processing filters
discussed in the [filtering vignette](Filtering.html).

```{r}
# ATAC
example_input$atac %>% print()
```

Segments have this format:

- `chr`, `from` and `to`: genomic range of the segment
- `copies`: ploidy of each segment

```{r}
example_input$segments %>% print()
```

# Creation of a new dataset

## Single-molecule assay (RNA)

Single-cell RNA, with z-score representation (Gaussian likelihood, `"G"`) for the RNA counts.

```{r}
x = init(
  rna = example_input$rna,
  atac = NULL,
  segmentation = example_input$segments,
  rna_likelihood = "G",
  description = 'SU008 sampled cells (example)')
```

S3 print method to visualise some information about the object

```{r}
x %>% print
```


<!-- ```{r, fig.width=6, fig.height=6} -->
<!-- # Select the first 5 segments to plot -->
<!-- some_segments = get_input(x, what = "segmentation") %>%  -->
<!--   filter(row_number() < 5) %>%  -->
<!--   pull(segment_id) -->

<!-- plot_data(x, what = 'histogram', segments = some_segments) -->
<!-- ``` -->

<!-- ```{r, fig.width=8, fig.height=4} -->
<!-- plot_data(x, what = 'lineplot') -->
<!-- ``` -->

<!-- ```{r, fig.width=6, fig.height=6} -->
<!-- plot_data(x, what = 'heatmap') -->
<!-- ``` -->


## Two-molecules assays (RNA and ATAC)

Joint single-cell RNA as above, but also single-cellATAC medeled with discrete likelihood for counts (Negative Binomial,  `"NB"`)

```{r}
x = init(
  rna = example_input$rna,
  atac = example_input$atac,
  segmentation = example_input$segments,
  rna_likelihood = "G", 
  atac_likelihood = 'NB', # This is a count-based modality
  description = 'SU008 sampled cells (example)')
```

The S3 print method reports the information about this object.

```{r}
x %>% print
```
<!-- ```{r, fig.width=6, fig.height=6} -->
<!-- plot_data(x, what = 'histogram', segments = some_segments) -->
<!-- ``` -->

<!-- ```{r, fig.width=8, fig.height=4} -->
<!-- plot_data(x, what = 'lineplot') -->
<!-- ``` -->

<!-- ```{r, fig.width=6, fig.height=8} -->
<!-- plot_data(x, what = 'heatmap') -->
<!-- ``` -->

# Datasets

## RNA

```{r}
# Breast-cancer PDX
data('campbell_bcpdx')
campbell_bcpdx %>% print
```

```{r}
# Glioblastoma tumour normal
data('patel_gbmtn')
patel_gbmtn %>% print
```

```{r}
# Hematological bone marrow failure
data('zaho_hemato')
zaho_hemato %>% print
```

## RNA and ATAC

```{r}
# Basal cell carcinoma
data('satpathy_bcc')
satpathy_bcc %>% print
```
