<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Glioblastoma smart-seq analysis • Rcongas</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Glioblastoma smart-seq analysis">
<meta property="og:description" content="Rcongas">
<meta property="og:image" content="militeee.github.io/Rcongas/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Rcongas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Getters.html">Getters</a>
    </li>
    <li>
      <a href="../articles/Plotting_functions.html">Plotting functions</a>
    </li>
    <li>
      <a href="../articles/gbm_example.html">Glioblastoma smart-seq analysis</a>
    </li>
    <li>
      <a href="../articles/monosomy_example.html">Monosomy 7 analysis</a>
    </li>
    <li>
      <a href="../articles/simulated_example.html">CONGAS on simulated data</a>
    </li>
    <li>
      <a href="../articles/xeno_examples.html">analysis on 10x data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/militeee/Rcongas/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="gbm_example_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Glioblastoma smart-seq analysis</h1>
                        <h4 class="author">Salvatore Milite</h4>
            
            <h4 class="date">2021-01-30</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/militeee/Rcongasvignettes/gbm_example.Rmd"><code>vignettes/gbm_example.Rmd</code></a></small>
      <div class="hidden name"><code>gbm_example.Rmd</code></div>

    </div>

    
    
<pre><code>#&gt; Loading Rcongas
#&gt; ✓ Loading Rcongas, 'Copy-Number genotyping from single cells'. Support : &lt;https://militeee.github.io/Rcongas/&gt;</code></pre>
<p>Here we present the analysis of a glioblastome smart-seq dataset from <span class="citation">(Patel et al. 2014)</span>. Count and allele frequencies matrices have been obtained from <span class="citation">(Fan et al. 2018)</span> and assembled in a set of .RData files.</p>
<p>We can first start by loading the data and trying to aggregate our gene counts over entire chromosomes. Even if the segmentation is not exactly precise, we can already note different potential subclonal CNV.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Rcongas</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"glioblastoma_TPM.rda"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"hg38_karyo"</span><span class="op">)</span>

<span class="va">chr_cnv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">hg38_karyo</span><span class="op">)</span>, start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">hg38_karyo</span><span class="op">)</span><span class="op">)</span>, end <span class="op">=</span> <span class="va">hg38_karyo</span>, tot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">hg38_karyo</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">express</span> <span class="op">&lt;-</span> <span class="fu">get_data</span><span class="op">(</span><span class="va">glioblastoma_TPM</span>, cnv_data <span class="op">=</span> <span class="va">chr_cnv</span>, type <span class="op">=</span> <span class="st">"fixed_binning"</span>, fun <span class="op">=</span> <span class="va">mean</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: biomaRt</span>

<span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">express</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span>, cluster_cols <span class="op">=</span> <span class="cn">F</span>,  show_rownames <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></code></pre></div>
<p><img src="gbm_example_files/figure-html/exploratory-1.png" width="700"></p>
<p>Some interesting chromosomes are the 9,14 and maybe 13. Now let’s segment the genome.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/">ggpubr</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: ggplot2</span>

<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"AF.rda"</span><span class="op">)</span>

<span class="co"># Basically we just convert the A and B allele information in major and minor</span>

<span class="va">cov_tr</span> <span class="op">&lt;-</span>  <span class="fl">20</span>

<span class="va">good_sites</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">AF_info</span><span class="op">$</span><span class="va">coverage</span> <span class="op">&gt;</span> <span class="va">cov_tr</span><span class="op">)</span>

<span class="va">MAF</span> <span class="op">&lt;-</span> <span class="va">AF_info</span><span class="op">$</span><span class="va">minor</span><span class="op">[</span><span class="va">good_sites</span><span class="op">]</span> <span class="op">/</span> <span class="va">AF_info</span><span class="op">$</span><span class="va">coverage</span><span class="op">[</span><span class="va">good_sites</span><span class="op">]</span>

<span class="va">MAF_l</span> <span class="op">&lt;-</span> <span class="va">MAF</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"pos"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html">separate</a></span><span class="op">(</span><span class="va">pos</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="st">"start"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># perform segmentation</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">segment_genome</span><span class="op">(</span><span class="va">MAF_l</span>, steps <span class="op">=</span> <span class="fl">200L</span>, param_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">'t'</span><span class="op">=</span><span class="fl">1e-8</span><span class="op">)</span>, filter_size <span class="op">=</span> <span class="fl">25</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">segs_df</span> <span class="op">&lt;-</span> <span class="fu">Rcongas</span><span class="fu">:::</span><span class="fu">from_inference_to_segments</span><span class="op">(</span><span class="va">res</span>, <span class="va">MAF_l</span><span class="op">)</span>

<span class="co"># to be honest I would not trust anything that is not an LOH from the output as the data is super variable</span>
<span class="va">segs_df</span><span class="op">$</span><span class="va">tot</span><span class="op">[</span><span class="va">segs_df</span><span class="op">$</span><span class="va">tot</span> <span class="op">==</span> <span class="fl">5</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">4</span>
<span class="va">segs_df</span><span class="op">$</span><span class="va">tot</span><span class="op">[</span><span class="va">segs_df</span><span class="op">$</span><span class="va">tot</span> <span class="op">==</span> <span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">4</span>
<span class="va">segs_df</span><span class="op">$</span><span class="va">tot</span><span class="op">[</span><span class="va">segs_df</span><span class="op">$</span><span class="va">tot</span> <span class="op">==</span> <span class="fl">6</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">4</span>

<span class="co"># collapse adjacent segments with the same CN value </span>
<span class="va">segs_df_corr</span> <span class="op">&lt;-</span> <span class="fu">Rcongas</span><span class="fu">:::</span><span class="fu">simplify_segs</span><span class="op">(</span><span class="va">segs_df</span><span class="op">)</span>

<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fu">plot_MAF</span><span class="op">(</span><span class="va">MAF_l</span>, k <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, <span class="fu">plot_MAF</span><span class="op">(</span><span class="va">MAF_l</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span>
 , align <span class="op">=</span> <span class="st">"hv"</span>, legend <span class="op">=</span> <span class="st">"none"</span>, font.label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="gbm_example_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># if you want you can also plot the hmm segments</span>
<span class="fu">plot_HMM</span><span class="op">(</span><span class="va">segs_df_corr</span><span class="op">)</span></code></pre></div>
<p><img src="gbm_example_files/figure-html/unnamed-chunk-5-2.png" width="700"></p>
<p>After that we can again aggregate counts over our new segmentation and have a look at the heatmap.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">segs_df_corr</span><span class="op">$</span><span class="va">chr</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="va">segs_df_corr</span><span class="op">$</span><span class="va">chr</span><span class="op">)</span>
<span class="co"># aggregate pver the chromosome</span>
<span class="va">binned_express</span> <span class="op">&lt;-</span> <span class="fu">get_data</span><span class="op">(</span><span class="va">glioblastoma_TPM</span>, cnv_data <span class="op">=</span> <span class="va">segs_df_corr</span>, type <span class="op">=</span> <span class="st">"fixed_binning"</span>, correct_bins <span class="op">=</span> <span class="cn">F</span>, fun <span class="op">=</span> <span class="va">mean</span>, bindim <span class="op">=</span> <span class="va">median</span>, genome <span class="op">=</span> <span class="st">"hg19"</span><span class="op">)</span>
<span class="co"># set a fake ploidy prior, it is needed </span>
<span class="va">binned_express</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">cnv</span><span class="op">$</span><span class="va">ploidy_real</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">binned_express</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">cnv</span><span class="op">)</span><span class="op">)</span>

<span class="va">binned_express</span> <span class="op">&lt;-</span> <span class="fu">filter_segments.rcongas</span><span class="op">(</span><span class="va">binned_express</span>, filter_mu <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>
<span class="co">#&gt;  [1]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE</span>
<span class="co">#&gt; [13]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE</span>
<span class="co">#&gt; [25]  TRUE  TRUE  TRUE</span>

<span class="co"># let's take note of our putative LOH</span>
<span class="va">LOH</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="va">binned_express</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">cnv</span><span class="op">$</span><span class="va">segment_id</span>, pattern <span class="op">=</span> <span class="st">"chr1[0|3|4]:"</span>, value <span class="op">=</span> <span class="cn">T</span>, perl <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>

<span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">binned_express</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span>, cluster_cols <span class="op">=</span> <span class="cn">F</span>,  show_rownames <span class="op">=</span> <span class="cn">F</span>, cluster_rows <span class="op">=</span> <span class="cn">T</span>, cutree_rows <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="gbm_example_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>Prior to performing the inference with SVI, we select just the more variable features, as high number of segments to cells ratio could led to an excessive clusters penalization.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>
<span class="va">torch</span> <span class="op">&lt;-</span> <span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/import.html">import</a></span><span class="op">(</span><span class="st">"torch"</span><span class="op">)</span>
<span class="co"># set priors on the mixing proportions</span>
<span class="va">mix_prior</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="cn">NULL</span>, <span class="va">torch</span><span class="op">$</span><span class="fu">tensor</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.8</span>,<span class="fl">0.2</span><span class="op">)</span><span class="op">)</span>, <span class="va">torch</span><span class="op">$</span><span class="fu">tensor</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.4</span>,<span class="fl">0.3</span>,<span class="fl">0.3</span><span class="op">)</span><span class="op">)</span>, <span class="va">torch</span><span class="op">$</span><span class="fu">tensor</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.3</span>,<span class="fl">0.3</span>,<span class="fl">0.2</span>,<span class="fl">0.2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co"># we filter the lessa variable segments</span>
<span class="va">filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">binned_express</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span>,<span class="fl">2</span>,<span class="va">sd</span><span class="op">)</span>
<span class="va">binned_express_most_var</span> <span class="op">&lt;-</span>  <span class="va">binned_express</span><span class="op">[</span>, <span class="va">filt</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">filt</span>, <span class="fl">0.70</span><span class="op">)</span><span class="op">]</span>
<span class="co"># run inference</span>
<span class="va">subclones</span> <span class="op">&lt;-</span> <span class="fu">best_cluster</span><span class="op">(</span>X <span class="op">=</span> <span class="va">binned_express_most_var</span>, clusters <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, model <span class="op">=</span> <span class="st">"MixtureGaussianNorm"</span>, steps <span class="op">=</span><span class="fl">800L</span>, lr <span class="op">=</span> <span class="fl">0.01</span>, param_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">0</span>, b <span class="op">=</span> <span class="fl">1</span>, init_sd <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"AIC"</span>, mixture <span class="op">=</span> <span class="va">mix_prior</span>, posteriors <span class="op">=</span> <span class="cn">T</span>, step_post <span class="op">=</span> <span class="fl">200</span><span class="op">)</span> </code></pre></div>
<p>CONGAS is able to identify 3 different clusters. We can clearly see how one of them can actually be a cluster of normal cells, as it lacks all our putative LOH regions.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#plot the results</span>
<span class="va">subclones</span><span class="op">$</span><span class="va">inference</span><span class="op">$</span><span class="va">models</span><span class="op">[[</span><span class="va">subclones</span><span class="op">$</span><span class="va">inference</span><span class="op">$</span><span class="va">model_selection</span><span class="op">$</span><span class="va">best_K</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">parameters</span><span class="op">$</span><span class="va">norm_factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">75</span><span class="op">)</span>
<span class="co">#subclones$data &lt;-  binned_express$data</span>
<span class="fu"><a href="../reference/plot_counts_rna_segments.html">plot_counts_rna_segments</a></span><span class="op">(</span><span class="va">subclones</span>, z_score <span class="op">=</span> <span class="cn">T</span>, normalised <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="co">#&gt; ✓ Loading CNAqc, 'Copy Number Alteration quality check'. Support : &lt;https://caravagn.github.io/CNAqc/&gt;</span></code></pre></div>
<p><img src="gbm_example_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>We can use this knoledge to filter the data and redo the segmentation just on tumor cells (which would be an ideal setting for our HMM).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># we now remove the normal cells and recalculate the major and minor allele proportions</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"MGH31_SNPs_per_cell.rda"</span><span class="op">)</span>

<span class="va">all_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu">get_cluster_assignments</span><span class="op">(</span><span class="va">subclones</span><span class="op">)</span><span class="op">)</span>

<span class="va">normal_cells</span> <span class="op">&lt;-</span> <span class="va">all_cells</span><span class="op">[</span><span class="fu">get_cluster_assignments</span><span class="op">(</span><span class="va">subclones</span><span class="op">)</span> <span class="op">==</span> <span class="st">"c3"</span><span class="op">]</span>
<span class="va">tumor_cells</span> <span class="op">&lt;-</span> <span class="va">all_cells</span><span class="op">[</span><span class="fu">get_cluster_assignments</span><span class="op">(</span><span class="va">subclones</span><span class="op">)</span> <span class="op">!=</span> <span class="st">"c3"</span><span class="op">]</span>


<span class="va">ref_normal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">MGH31_SNPs_per_cell</span><span class="op">$</span><span class="va">ref</span><span class="op">[</span>,<span class="va">normal_cells</span><span class="op">]</span><span class="op">)</span>
<span class="va">alt_normal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">MGH31_SNPs_per_cell</span><span class="op">$</span><span class="va">alt</span><span class="op">[</span>,<span class="va">normal_cells</span><span class="op">]</span><span class="op">)</span>
<span class="va">min_normal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">pmin</a></span><span class="op">(</span><span class="va">ref_normal</span>, <span class="va">alt_normal</span><span class="op">)</span>
<span class="va">maj_normal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">pmax</a></span><span class="op">(</span><span class="va">ref_normal</span>, <span class="va">alt_normal</span><span class="op">)</span>
<span class="va">cov_normal</span> <span class="op">&lt;-</span> <span class="va">ref_normal</span> <span class="op">+</span> <span class="va">alt_normal</span>

<span class="va">ref_tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">MGH31_SNPs_per_cell</span><span class="op">$</span><span class="va">ref</span><span class="op">[</span>,<span class="va">tumor_cells</span><span class="op">]</span><span class="op">)</span>
<span class="va">alt_tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">MGH31_SNPs_per_cell</span><span class="op">$</span><span class="va">alt</span><span class="op">[</span>,<span class="va">tumor_cells</span><span class="op">]</span><span class="op">)</span>
<span class="va">min_tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">pmin</a></span><span class="op">(</span><span class="va">ref_tumor</span>, <span class="va">alt_tumor</span><span class="op">)</span>
<span class="va">maj_tumor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">pmax</a></span><span class="op">(</span><span class="va">ref_tumor</span>, <span class="va">alt_tumor</span><span class="op">)</span>
<span class="va">cov_tumor</span> <span class="op">&lt;-</span> <span class="va">ref_tumor</span> <span class="op">+</span> <span class="va">alt_tumor</span>

<span class="va">filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cov_normal</span> <span class="op">&gt;</span> <span class="fl">30</span><span class="op">)</span>

<span class="va">MAF_normal</span> <span class="op">&lt;-</span> <span class="va">min_normal</span><span class="op">[</span><span class="va">filter</span><span class="op">]</span> <span class="op">/</span> <span class="va">cov_normal</span><span class="op">[</span><span class="va">filter</span><span class="op">]</span>
<span class="va">MAF_tumor</span> <span class="op">&lt;-</span> <span class="va">min_tumor</span><span class="op">[</span><span class="va">filter</span><span class="op">]</span> <span class="op">/</span> <span class="va">cov_tumor</span><span class="op">[</span><span class="va">filter</span><span class="op">]</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># basically the same thing as before but with just tumor cells</span>

<span class="va">cov_tr</span> <span class="op">&lt;-</span>  <span class="fl">40</span>

<span class="va">good_sites</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cov_tumor</span> <span class="op">&gt;</span> <span class="va">cov_tr</span><span class="op">)</span>

<span class="va">MAF</span> <span class="op">&lt;-</span> <span class="va">min_tumor</span><span class="op">[</span><span class="va">good_sites</span><span class="op">]</span> <span class="op">/</span> <span class="va">cov_tumor</span><span class="op">[</span><span class="va">good_sites</span><span class="op">]</span>

<span class="va">MAF</span> <span class="op">&lt;-</span>  <span class="va">MAF</span><span class="op">[</span><span class="va">MAF</span> <span class="op">&gt;</span> <span class="fl">0.10</span><span class="op">]</span>

<span class="va">MAF</span> <span class="op">&lt;-</span>  <span class="va">MAF</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">MAF</span><span class="op">)</span>, pattern <span class="op">=</span> <span class="st">"chr1[0|4|3]:"</span>,invert <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">]</span>

<span class="va">MAF_l</span> <span class="op">&lt;-</span> <span class="va">MAF</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"pos"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html">separate</a></span><span class="op">(</span><span class="va">pos</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="st">"start"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">segment_genome</span><span class="op">(</span><span class="va">MAF_l</span>, steps <span class="op">=</span> <span class="fl">200L</span>, param_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">'t'</span> <span class="op">=</span><span class="fl">1e-8</span><span class="op">)</span>, filter_size <span class="op">=</span> <span class="fl">33</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">segs_df</span> <span class="op">&lt;-</span> <span class="fu">from_inference_to_segments</span><span class="op">(</span><span class="va">res</span>, <span class="va">MAF_l</span><span class="op">)</span>

<span class="va">segs_df</span><span class="op">$</span><span class="va">tot</span><span class="op">[</span><span class="va">segs_df</span><span class="op">$</span><span class="va">tot</span> <span class="op">==</span> <span class="fl">6</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">2</span>

<span class="va">segs_df</span><span class="op">$</span><span class="va">tot</span><span class="op">[</span><span class="va">segs_df</span><span class="op">$</span><span class="va">tot</span> <span class="op">==</span> <span class="fl">5</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">4</span>

<span class="va">segs_df_corr2</span> <span class="op">&lt;-</span> <span class="fu">simplify_segs</span><span class="op">(</span><span class="va">segs_df</span><span class="op">)</span>


<span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span><span class="fu">plot_MAF</span><span class="op">(</span><span class="va">MAF_l</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>, <span class="fu">plot_MAF</span><span class="op">(</span><span class="va">MAF_l</span>, k <span class="op">=</span> <span class="fl">11</span><span class="op">)</span>, align <span class="op">=</span> <span class="st">"hv"</span>, common.legend <span class="op">=</span> <span class="cn">T</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, legend <span class="op">=</span> <span class="st">"none"</span>, font.label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Warning in runmed(x$value, k = k, endrule = "constant"): 'k' is bigger than 'n'!</span>
<span class="co">#&gt; Changing 'k' to 3</span>
<span class="co">#&gt; Warning in runmed(x$value, k = k, endrule = "constant"): 'k' is bigger than 'n'!</span>
<span class="co">#&gt; Changing 'k' to 1</span>
<span class="co">#&gt; Warning in runmed(x$value, k = k, endrule = "constant"): 'k' is bigger than 'n'!</span>
<span class="co">#&gt; Changing 'k' to 3</span>
<span class="co">#&gt; Warning in runmed(x$value, k = k, endrule = "constant"): 'k' is bigger than 'n'!</span>
<span class="co">#&gt; Changing 'k' to 1</span>
<span class="co">#&gt; Warning in runmed(x$value, k = k, endrule = "constant"): 'k' is bigger than 'n'!</span>
<span class="co">#&gt; Changing 'k' to 7</span></code></pre></div>
<p><img src="gbm_example_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>We can now aggregate expression values on the new segments.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">segs_df_corr2</span><span class="op">$</span><span class="va">chr</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="va">segs_df_corr2</span><span class="op">$</span><span class="va">chr</span><span class="op">)</span>

<span class="va">binned_express2</span> <span class="op">&lt;-</span> <span class="fu">get_data</span><span class="op">(</span><span class="va">glioblastoma_TPM</span><span class="op">[</span>,<span class="va">tumor_cells</span><span class="op">]</span>, cnv_data <span class="op">=</span> <span class="va">segs_df_corr2</span>, type <span class="op">=</span> <span class="st">"fixed_binning"</span>, correct_bins <span class="op">=</span> <span class="cn">F</span>, fun <span class="op">=</span> <span class="va">mean</span>, chrs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">11</span><span class="op">]</span><span class="op">)</span>

<span class="va">binned_express2</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">cnv</span><span class="op">$</span><span class="va">ploidy_real</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">binned_express2</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">cnv</span><span class="op">)</span><span class="op">)</span>

<span class="va">binned_express2</span> <span class="op">&lt;-</span> <span class="fu">filter_segments.rcongas</span><span class="op">(</span><span class="va">binned_express2</span>, filter_mu <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>
<span class="co"># manually add the counts from the LOH region aggregated before, as they are lost in this tumor-only segmentation process</span>
<span class="va">final_expr</span> <span class="op">&lt;-</span> <span class="fu">bind_rcongas_data</span><span class="op">(</span><span class="va">binned_express2</span>, <span class="va">binned_express</span><span class="op">[</span><span class="va">tumor_cells</span>,<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">binned_express</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>  <span class="op">%in%</span> <span class="va">LOH</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">final_expr</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span>, cluster_cols <span class="op">=</span> <span class="cn">F</span>,  show_rownames <span class="op">=</span> <span class="cn">F</span>, cluster_rows <span class="op">=</span> <span class="cn">T</span>, cutree_rows <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="gbm_example_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>We filter again for low variance segments and run CONGAS.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>
<span class="co"># fit CONGAS with just tumor cells </span>
<span class="va">to_run</span> <span class="op">&lt;-</span> <span class="va">final_expr</span><span class="op">[</span><span class="va">tumor_cells</span>,<span class="op">]</span>

<span class="va">filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">to_run</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span>,<span class="fl">2</span>,<span class="va">sd</span><span class="op">)</span>

<span class="va">to_run</span> <span class="op">&lt;-</span>  <span class="va">to_run</span><span class="op">[</span>,<span class="va">filt</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">filt</span>, <span class="fl">0.70</span><span class="op">)</span><span class="op">]</span>

<span class="va">subclones2</span> <span class="op">&lt;-</span> <span class="fu">best_cluster</span><span class="op">(</span><span class="va">to_run</span>, clusters <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, model <span class="op">=</span> <span class="st">"MixtureGaussianNorm"</span>, steps <span class="op">=</span> <span class="fl">800L</span>, lr <span class="op">=</span> <span class="fl">0.01</span>, param_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">0</span>, b <span class="op">=</span> <span class="fl">1</span>, init_sd <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"lk"</span>, posteriors <span class="op">=</span> <span class="cn">T</span>, step_post <span class="op">=</span> <span class="fl">150</span><span class="op">)</span> </code></pre></div>
<p>Our new 3 clusters correspond to different clonal populations: a parental one, one with a deletion at 14 and one with an amplification on chromosome 5. With this low number of cells however, differential expression analysis has not enough power.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_counts_rna_segments.html">plot_counts_rna_segments</a></span><span class="op">(</span><span class="va">subclones2</span>, z_score <span class="op">=</span> <span class="cn">T</span>, normalised <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></code></pre></div>
<p><img src="gbm_example_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#clone specific differential expression</span>

<span class="va">df_norm_tum</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/calculate_DE.html">calculate_DE</a></span><span class="op">(</span><span class="va">subclones</span>, input <span class="op">=</span> <span class="va">glioblastoma_TPM</span>, clone1 <span class="op">=</span> <span class="st">"c3"</span>, <span class="cn">NULL</span>, normalize <span class="op">=</span> <span class="cn">F</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>
  
<span class="va">df_5_no5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_DE.html">calculate_DE</a></span><span class="op">(</span><span class="va">subclones2</span>, input <span class="op">=</span> <span class="va">glioblastoma_TPM</span><span class="op">[</span>,<span class="fu">get_best_model</span><span class="op">(</span><span class="va">subclones2</span><span class="op">)</span><span class="op">$</span><span class="va">dim_names</span><span class="op">$</span><span class="va">cell_names</span><span class="op">]</span>, clone1 <span class="op">=</span> <span class="st">"c2"</span>, <span class="cn">NULL</span>, normalize <span class="op">=</span> <span class="cn">F</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>
  
<span class="va">df_14_no14</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_DE.html">calculate_DE</a></span><span class="op">(</span><span class="va">subclones2</span>, input <span class="op">=</span> <span class="va">glioblastoma_TPM</span><span class="op">[</span>,<span class="fu">get_best_model</span><span class="op">(</span><span class="va">subclones2</span><span class="op">)</span><span class="op">$</span><span class="va">dim_names</span><span class="op">$</span><span class="va">cell_names</span><span class="op">]</span>, clone1 <span class="op">=</span> <span class="st">"c1"</span>,<span class="st">"c3"</span>, normalize <span class="op">=</span> <span class="cn">F</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-hb" class="csl-entry">
Fan, Jean, Hae-Ock Lee, Soohyun Lee, Da-eun Ryu, Semin Lee, Catherine Xue, Seok Jin Kim, et al. 2018. <span>“Linking Transcriptional and Genetic Tumor Heterogeneity Through Allele Analysis of Single-Cell RNA-Seq Data.”</span> <em>Genome Research</em> 28 (8): 1217–27.
</div>
<div id="ref-patel" class="csl-entry">
Patel, Anoop P, Itay Tirosh, John J Trombetta, Alex K Shalek, Shawn M Gillespie, Hiroaki Wakimoto, Daniel P Cahill, et al. 2014. <span>“Single-Cell RNA-Seq Highlights Intratumoral Heterogeneity in Primary Glioblastoma.”</span> <em>Science</em> 344 (6190): 1396–1401.
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://sites.google.com/view/cdslab">Salvatore Milite</a>, <a href="https://sites.google.com/view/cdslab">Giulio Caravagna</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
