<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getters • Rcongas</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Getters">
<meta property="og:description" content="Rcongas">
<meta property="og:image" content="militeee.github.io/Rcongas/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Rcongas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Getters.html">Getters</a>
    </li>
    <li>
      <a href="../articles/Plotting_functions.html">Plotting functions</a>
    </li>
    <li>
      <a href="../articles/simulated_example.html">CONFETTI on simulated data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/militeee/Rcongas/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Getters_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getters</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/militeee/Rcongasvignettes/Getters.Rmd"><code>vignettes/Getters.Rmd</code></a></small>
      <div class="hidden name"><code>Getters.Rmd</code></div>

    </div>

    
    
<p>We show how you can access information from a CONGAS computation, using the <code>rcongas</code> package.</p>
<div id="getters-available-in-rcongas" class="section level1">
<h1 class="hasAnchor">
<a href="#getters-available-in-rcongas" class="anchor"></a>Getters available in <code>rcongas</code>
</h1>
<p>We work with an example object released with the tool.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co"># change to data('rcongas_example', package = 'rcongas')</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"../data/rcongas_example.rda"</span><span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: tidyverse</span>
<span class="co">#&gt; ── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──</span>
<span class="co">#&gt; ✓ ggplot2 3.3.2     ✓ purrr   0.3.4</span>
<span class="co">#&gt; ✓ tibble  3.0.3     ✓ dplyr   1.0.2</span>
<span class="co">#&gt; ✓ tidyr   1.1.2     ✓ stringr 1.4.0</span>
<span class="co">#&gt; ✓ readr   1.4.0     ✓ forcats 0.5.0</span>
<span class="co">#&gt; ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──</span>
<span class="co">#&gt; x dplyr::filter() masks stats::filter()</span>
<span class="co">#&gt; x dplyr::lag()    masks stats::lag()</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">Rcongas</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: Rcongas</span>
<span class="co">#&gt; ✓ Loading Rcongas, 'Copy-Number genotyping from single cells'. Support : &lt;https://militeee.github.io/Rcongas/&gt;</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'Rcongas'</span>
<span class="co">#&gt; The following object is masked _by_ '.GlobalEnv':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     rcongas_example</span>

<span class="va">rcongas_example</span> <span class="op">=</span> <span class="fu">Rcongas</span><span class="fu">::</span><span class="va"><a href="../reference/rcongas_example.html">rcongas_example</a></span>

<span class="co"># sapply(list.files('../R/', full.names = T), source)</span>
<span class="co"># source('../R/getters_clustering_results.R')</span>
<span class="co"># source('../R/getters_extra_analysis.R')</span>
<span class="co"># source('../R/getters_input_data.R')</span>
<span class="co"># source('../R/getters_privates.R')</span>
<span class="co"># source('../R/utils.R')</span>
<span class="co"># source('../R/test_status.R')</span></pre></div>
<div id="reference-information" class="section level2">
<h2 class="hasAnchor">
<a href="#reference-information" class="anchor"></a>Reference information</h2>
<ul>
<li>
<code>get_reference_genome</code>, <code>get_gene_annotations</code> to obtain information about the used reference.</li>
</ul>
<p>The gene locations for the reference genomes supported by congas can be obtained by using <code>get_gene_annotations</code>.</p>
<p>Note that the reference genome id is stored in the <code>rcongas</code> object created at the time the analysis is performed. So the genome coordinates of the genes are consistent with the adopted reference.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co"># The adopted reference genome</span>
<span class="fu"><a href="../reference/get_reference_genome.html">get_reference_genome</a></span><span class="op">(</span><span class="va">rcongas_example</span><span class="op">)</span>

<span class="co"># Gene coordinates as of the reference</span>
<span class="fu"><a href="../reference/get_gene_annotations.html">get_gene_annotations</a></span><span class="op">(</span><span class="va">rcongas_example</span><span class="op">)</span></pre></div>
</div>
<div id="clustering-results" class="section level2">
<h2 class="hasAnchor">
<a href="#clustering-results" class="anchor"></a>Clustering results</h2>
<ul>
<li>
<code>get_k</code>, <code>get_clusters_size</code>, <code>get_clones_ploidy</code> to obtain information about the output clusters;</li>
<li>
<code>get_clusters</code> to obtain clustering assignments.</li>
</ul>
<p>You can obtain the value fit for <span class="math inline">\(k\geq 1\)</span>, the number of clusters/ clones in the CONGAS model, using <code>get_k</code>. For every cluster, you can also obtain the total number of cells assigned to that, as well as its normalised value (i.e., the mixing proporitions).</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="co"># Best number of clusters/ clones in the CONGAS model fit</span>
<span class="fu"><a href="../reference/get_k.html">get_k</a></span><span class="op">(</span><span class="va">rcongas_example</span><span class="op">)</span>
<span class="co">#&gt; [1] 2</span>

<span class="co"># Cluster size</span>
<span class="fu"><a href="../reference/get_clusters_size.html">get_clusters_size</a></span><span class="op">(</span><span class="va">rcongas_example</span><span class="op">)</span>
<span class="co">#&gt;   1   2 </span>
<span class="co">#&gt; 369 135</span>

<span class="co"># Cluster size (normalised)</span>
<span class="fu"><a href="../reference/get_clusters_size.html">get_clusters_size</a></span><span class="op">(</span><span class="va">rcongas_example</span>, normalise <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt;         1         2 </span>
<span class="co">#&gt; 0.7321429 0.2678571</span></pre></div>
<p>Then you can also obtain the CNA profile estimated for each one of the <span class="math inline">\(k\)</span> clones, using <code>get_clones_ploidy</code>. This reports, for every segment the lognormal CNA value estimated by CONGAS, as well as the segment mean, and the <code>CN</code> value (which is the offset to the mean of the segment). Values from this getter can be automatically filtered for certain chromsome ids, and clusters.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="co"># Get all data (all clones, all chromsomes)</span>
<span class="fu"><a href="../reference/get_clones_ploidy.html">get_clones_ploidy</a></span><span class="op">(</span><span class="va">rcongas_example</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 56 x 7</span>
<span class="co">#&gt;    chr        from        to cluster       CN segment_mean lognorm_CN</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 chr1   12600001 156000000 1       -0.00547         1.27       1.26</span>
<span class="co">#&gt;  2 chr1  156000001 249300000 1        0.00502         1.20       1.21</span>
<span class="co">#&gt;  3 chr2   26250001 159900000 1       -0.00468         1.21       1.20</span>
<span class="co">#&gt;  4 chr2  176850001 243300000 1        0.00283         1.17       1.17</span>
<span class="co">#&gt;  5 chr3   34350001  86850000 1       -0.0163          1.07       1.05</span>
<span class="co">#&gt;  6 chr3  114000001 152100000 1        0.0197          1.28       1.30</span>
<span class="co">#&gt;  7 chr4   70500001 135300000 1        0.0193          1.17       1.19</span>
<span class="co">#&gt;  8 chr5   54150001 127050000 1       -0.00631         1.10       1.09</span>
<span class="co">#&gt;  9 chr5  129150001 150600000 1       -0.0201          1.32       1.30</span>
<span class="co">#&gt; 10 chr6   14850001  32400000 1       -0.0512          1.35       1.30</span>
<span class="co">#&gt; # … with 46 more rows</span>

<span class="co"># Zoom in one specific chromosome</span>
<span class="fu"><a href="../reference/get_clones_ploidy.html">get_clones_ploidy</a></span><span class="op">(</span><span class="va">rcongas_example</span>, chromosomes <span class="op">=</span> <span class="st">'chr2'</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 4 x 7</span>
<span class="co">#&gt;   chr        from        to cluster       CN segment_mean lognorm_CN</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt; 1 chr2   26250001 159900000 1       -0.00468         1.21       1.20</span>
<span class="co">#&gt; 2 chr2  176850001 243300000 1        0.00283         1.17       1.17</span>
<span class="co">#&gt; 3 chr2   26250001 159900000 2        0.00468         1.21       1.21</span>
<span class="co">#&gt; 4 chr2  176850001 243300000 2       -0.00283         1.17       1.17</span>

<span class="co"># Zoom in one specific cluster, here "1"</span>
<span class="fu"><a href="../reference/get_clones_ploidy.html">get_clones_ploidy</a></span><span class="op">(</span><span class="va">rcongas_example</span>, clusters <span class="op">=</span> <span class="st">'1'</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 28 x 7</span>
<span class="co">#&gt;    chr        from        to cluster       CN segment_mean lognorm_CN</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 chr1   12600001 156000000 1       -0.00547         1.27       1.26</span>
<span class="co">#&gt;  2 chr1  156000001 249300000 1        0.00502         1.20       1.21</span>
<span class="co">#&gt;  3 chr2   26250001 159900000 1       -0.00468         1.21       1.20</span>
<span class="co">#&gt;  4 chr2  176850001 243300000 1        0.00283         1.17       1.17</span>
<span class="co">#&gt;  5 chr3   34350001  86850000 1       -0.0163          1.07       1.05</span>
<span class="co">#&gt;  6 chr3  114000001 152100000 1        0.0197          1.28       1.30</span>
<span class="co">#&gt;  7 chr4   70500001 135300000 1        0.0193          1.17       1.19</span>
<span class="co">#&gt;  8 chr5   54150001 127050000 1       -0.00631         1.10       1.09</span>
<span class="co">#&gt;  9 chr5  129150001 150600000 1       -0.0201          1.32       1.30</span>
<span class="co">#&gt; 10 chr6   14850001  32400000 1       -0.0512          1.35       1.30</span>
<span class="co">#&gt; # … with 18 more rows</span>

<span class="co"># Zoom in one specific chromosome and cluster</span>
<span class="fu"><a href="../reference/get_clones_ploidy.html">get_clones_ploidy</a></span><span class="op">(</span><span class="va">rcongas_example</span>, chromosomes <span class="op">=</span> <span class="st">'chr2'</span>, clusters <span class="op">=</span> <span class="st">"1"</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 2 x 7</span>
<span class="co">#&gt;   chr        from        to cluster       CN segment_mean lognorm_CN</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt; 1 chr2   26250001 159900000 1       -0.00468         1.21       1.20</span>
<span class="co">#&gt; 2 chr2  176850001 243300000 1        0.00283         1.17       1.17</span></pre></div>
<p>Then you can also obtain the clustering assignments for each one of the input cells, using function <code>get_clusters</code>. This getter can also subset cells in order to contain only those that belong to a specific set of clusters.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co"># Get all cells at once</span>
<span class="fu"><a href="../reference/get_clusters.html">get_clusters</a></span><span class="op">(</span><span class="va">rcongas_example</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 504 x 2</span>
<span class="co">#&gt;    cell               cluster</span>
<span class="co">#&gt;    &lt;chr&gt;              &lt;chr&gt;  </span>
<span class="co">#&gt;  1 AAACCTGTCGACAGCC-1 2      </span>
<span class="co">#&gt;  2 AAAGATGGTAATAGCA-1 1      </span>
<span class="co">#&gt;  3 AACACGTAGCTATGCT-1 1      </span>
<span class="co">#&gt;  4 AACTCCCAGACAGGCT-1 2      </span>
<span class="co">#&gt;  5 AACTCCCGTACACCGC-1 1      </span>
<span class="co">#&gt;  6 AACTGGTCAAAGAATC-1 2      </span>
<span class="co">#&gt;  7 AACTGGTCAAAGTGCG-1 1      </span>
<span class="co">#&gt;  8 AACTTTCGTGGCTCCA-1 2      </span>
<span class="co">#&gt;  9 AACTTTCTCGGTTAAC-1 2      </span>
<span class="co">#&gt; 10 AAGACCTAGCCAGTAG-1 2      </span>
<span class="co">#&gt; # … with 494 more rows</span>

<span class="co"># Subset to cluster "1"</span>
<span class="fu"><a href="../reference/get_clusters.html">get_clusters</a></span><span class="op">(</span><span class="va">rcongas_example</span>, clusters <span class="op">=</span> <span class="st">'1'</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 369 x 2</span>
<span class="co">#&gt;    cell               cluster</span>
<span class="co">#&gt;    &lt;chr&gt;              &lt;chr&gt;  </span>
<span class="co">#&gt;  1 AAAGATGGTAATAGCA-1 1      </span>
<span class="co">#&gt;  2 AACACGTAGCTATGCT-1 1      </span>
<span class="co">#&gt;  3 AACTCCCGTACACCGC-1 1      </span>
<span class="co">#&gt;  4 AACTGGTCAAAGTGCG-1 1      </span>
<span class="co">#&gt;  5 AAGCCGCGTTGTACAC-1 1      </span>
<span class="co">#&gt;  6 AATCCAGCATTTCAGG-1 1      </span>
<span class="co">#&gt;  7 AATCCAGTCCGAACGC-1 1      </span>
<span class="co">#&gt;  8 AATCGGTAGTGTTGAA-1 1      </span>
<span class="co">#&gt;  9 AATCGGTTCTGTCTAT-1 1      </span>
<span class="co">#&gt; 10 ACACCAACAGAAGCAC-1 1      </span>
<span class="co">#&gt; # … with 359 more rows</span></pre></div>
</div>
<div id="stored-input-data" class="section level2">
<h2 class="hasAnchor">
<a href="#stored-input-data" class="anchor"></a>Stored input data</h2>
<ul>
<li>
<code>get_input_segmentation</code> to obtain information about the input DNA segmentation;</li>
<li>
<code>get_input_raw_data</code> and <code>get_counts</code> to obtain the raw input data (i.e., single-cell counts) and the counts per segment used by CONGAS.</li>
</ul>
<p>The input DNA segmentation is stored in the result of a CONGAS run. You can access that with <code>get_input_segmentation</code> and create simple filters based on chromosome ids.</p>
<p>Note that this tibble contains also columns that report the ploidy of the input segment (<code>ploidy_real</code>), the actual number of used genes retained that map to each segment (<code>mu</code>), and the total number of genes that map to the segments (<code>fixed_mu</code>).</p>
<p>Genes mapped to segments are filtered as explained in <code>XXXX</code>. Column <code>segment_id</code> is just a key to identify univocally the input segments to CONGAS.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="co"># All the input segments</span>
<span class="fu"><a href="../reference/get_input_segmentation.html">get_input_segmentation</a></span><span class="op">(</span><span class="va">rcongas_example</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 28 x 7</span>
<span class="co">#&gt;         size ploidy_real    mu fixed_mu chr        from        to</span>
<span class="co">#&gt;        &lt;dbl&gt;       &lt;int&gt; &lt;dbl&gt;    &lt;int&gt; &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt;  1 143399999           2  202.      603 chr1   12600001 156000000</span>
<span class="co">#&gt;  2  93299999           3  116       350 chr1  156000001 249300000</span>
<span class="co">#&gt;  3 133649999           2  130.      386 chr2   26250001 159900000</span>
<span class="co">#&gt;  4  66449999           2   61       192 chr2  176850001 243300000</span>
<span class="co">#&gt;  5  52499999           1   41       165 chr3   34350001  86850000</span>
<span class="co">#&gt;  6  38099999           3   51       134 chr3  114000001 152100000</span>
<span class="co">#&gt;  7  64799999           2   39       128 chr4   70500001 135300000</span>
<span class="co">#&gt;  8  72899999           1   37       154 chr5   54150001 127050000</span>
<span class="co">#&gt;  9  21449999           2   35        98 chr5  129150001 150600000</span>
<span class="co">#&gt; 10  17549999           3   44       126 chr6   14850001  32400000</span>
<span class="co">#&gt; # … with 18 more rows</span>

<span class="co"># Segments in chromosomes 1 to 3 (ids: chr1, chr2, chr3)</span>
<span class="fu"><a href="../reference/get_input_segmentation.html">get_input_segmentation</a></span><span class="op">(</span><span class="va">rcongas_example</span>, chromosomes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'chr'</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 x 7</span>
<span class="co">#&gt;        size ploidy_real    mu fixed_mu chr        from        to</span>
<span class="co">#&gt;       &lt;dbl&gt;       &lt;int&gt; &lt;dbl&gt;    &lt;int&gt; &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1 143399999           2  202.      603 chr1   12600001 156000000</span>
<span class="co">#&gt; 2  93299999           3  116       350 chr1  156000001 249300000</span>
<span class="co">#&gt; 3 133649999           2  130.      386 chr2   26250001 159900000</span>
<span class="co">#&gt; 4  66449999           2   61       192 chr2  176850001 243300000</span>
<span class="co">#&gt; 5  52499999           1   41       165 chr3   34350001  86850000</span>
<span class="co">#&gt; 6  38099999           3   51       134 chr3  114000001 152100000</span></pre></div>
<p>If you have run the main <code>rcongas</code> fit function <code>XXX</code> by using parameters <code>XXXX</code>, the raw input RNA data matrix have also been stored into the fit object (otherwise no, just to save space in the final object).</p>
<p>You have a getter <code>get_input_raw_data</code> to obtain the input data when this has been stored in a <code>rcongas</code> object, or you need to load yourself the input data that you used to fit the model (in that case <code>get_input_raw_data</code> would return <code>NULL</code> and report a warning to screen). The raw input data are the counts, per gene (rows), per cell (columns), obtained from the scRNAseq experiment.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="co"># Input data - stored inside "rcongas_example" in this case</span>
<span class="va">raw_input_gene_by_cell</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_input_raw_data.html">get_input_raw_data</a></span><span class="op">(</span><span class="va">rcongas_example</span><span class="op">)</span>

<span class="co"># Just the 2x2 submatrix (genes on the rows, cells on the columns)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">raw_input_gene_by_cell</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;          AAACCTGTCGACAGCC-1 AAAGATGGTAATAGCA-1</span>
<span class="co">#&gt; DHRS3                     0                  0</span>
<span class="co">#&gt; HNRNPCL1                  0                  0</span></pre></div>
<p>The data processed by CONGAS are counts mapped to segments, which are stored in a matrix that contains cells on rows, and segments on columns. This matrix can be extracted in long tidy format; the actual counts value can be obtained before or after normalisation, and can be subset to certain portions of the cancer genome.</p>
<p>In this long view we report also the cluster id for each of the input cells.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="co"># Normalised data (default)</span>
<span class="fu"><a href="../reference/get_counts.html">get_counts</a></span><span class="op">(</span><span class="va">rcongas_example</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 14,112 x 6</span>
<span class="co">#&gt;    chr       from        to cell                   n cluster</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;     &lt;int&gt; &lt;chr&gt;              &lt;dbl&gt; &lt;chr&gt;  </span>
<span class="co">#&gt;  1 chr1  12600001 156000000 AAACCTGTCGACAGCC-1  257. 2      </span>
<span class="co">#&gt;  2 chr1  12600001 156000000 AAAGATGGTAATAGCA-1  262. 1      </span>
<span class="co">#&gt;  3 chr1  12600001 156000000 AACACGTAGCTATGCT-1  242. 1      </span>
<span class="co">#&gt;  4 chr1  12600001 156000000 AACTCCCAGACAGGCT-1  253. 2      </span>
<span class="co">#&gt;  5 chr1  12600001 156000000 AACTCCCGTACACCGC-1  257. 1      </span>
<span class="co">#&gt;  6 chr1  12600001 156000000 AACTGGTCAAAGAATC-1  220. 2      </span>
<span class="co">#&gt;  7 chr1  12600001 156000000 AACTGGTCAAAGTGCG-1  234. 1      </span>
<span class="co">#&gt;  8 chr1  12600001 156000000 AACTTTCGTGGCTCCA-1  232. 2      </span>
<span class="co">#&gt;  9 chr1  12600001 156000000 AACTTTCTCGGTTAAC-1  265. 2      </span>
<span class="co">#&gt; 10 chr1  12600001 156000000 AAGACCTAGCCAGTAG-1  247. 2      </span>
<span class="co">#&gt; # … with 14,102 more rows</span>

<span class="co"># A view on a certain chromosome </span>
<span class="fu"><a href="../reference/get_counts.html">get_counts</a></span><span class="op">(</span><span class="va">rcongas_example</span>,
           chromosomes <span class="op">=</span> <span class="st">'chr3'</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 1,008 x 6</span>
<span class="co">#&gt;    chr       from       to cell                   n cluster</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt; &lt;chr&gt;              &lt;dbl&gt; &lt;chr&gt;  </span>
<span class="co">#&gt;  1 chr3  34350001 86850000 AAACCTGTCGACAGCC-1  45.5 2      </span>
<span class="co">#&gt;  2 chr3  34350001 86850000 AAAGATGGTAATAGCA-1  42.7 1      </span>
<span class="co">#&gt;  3 chr3  34350001 86850000 AACACGTAGCTATGCT-1  46.3 1      </span>
<span class="co">#&gt;  4 chr3  34350001 86850000 AACTCCCAGACAGGCT-1  36.6 2      </span>
<span class="co">#&gt;  5 chr3  34350001 86850000 AACTCCCGTACACCGC-1  32.9 1      </span>
<span class="co">#&gt;  6 chr3  34350001 86850000 AACTGGTCAAAGAATC-1  54.7 2      </span>
<span class="co">#&gt;  7 chr3  34350001 86850000 AACTGGTCAAAGTGCG-1  35.9 1      </span>
<span class="co">#&gt;  8 chr3  34350001 86850000 AACTTTCGTGGCTCCA-1  37.7 2      </span>
<span class="co">#&gt;  9 chr3  34350001 86850000 AACTTTCTCGGTTAAC-1  43.7 2      </span>
<span class="co">#&gt; 10 chr3  34350001 86850000 AAGACCTAGCCAGTAG-1  35.6 2      </span>
<span class="co">#&gt; # … with 998 more rows</span>

<span class="co"># Un-normalised data</span>
<span class="fu"><a href="../reference/get_counts.html">get_counts</a></span><span class="op">(</span><span class="va">rcongas_example</span>, normalise <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 14,112 x 6</span>
<span class="co">#&gt;    chr       from        to cell                   n cluster</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;     &lt;int&gt; &lt;chr&gt;              &lt;dbl&gt; &lt;chr&gt;  </span>
<span class="co">#&gt;  1 chr1  12600001 156000000 AAACCTGTCGACAGCC-1   322 2      </span>
<span class="co">#&gt;  2 chr1  12600001 156000000 AAAGATGGTAATAGCA-1   405 1      </span>
<span class="co">#&gt;  3 chr1  12600001 156000000 AACACGTAGCTATGCT-1   293 1      </span>
<span class="co">#&gt;  4 chr1  12600001 156000000 AACTCCCAGACAGGCT-1   339 2      </span>
<span class="co">#&gt;  5 chr1  12600001 156000000 AACTCCCGTACACCGC-1   312 1      </span>
<span class="co">#&gt;  6 chr1  12600001 156000000 AACTGGTCAAAGAATC-1   329 2      </span>
<span class="co">#&gt;  7 chr1  12600001 156000000 AACTGGTCAAAGTGCG-1   254 1      </span>
<span class="co">#&gt;  8 chr1  12600001 156000000 AACTTTCGTGGCTCCA-1   382 2      </span>
<span class="co">#&gt;  9 chr1  12600001 156000000 AACTTTCTCGGTTAAC-1   389 2      </span>
<span class="co">#&gt; 10 chr1  12600001 156000000 AAGACCTAGCCAGTAG-1   312 2      </span>
<span class="co">#&gt; # … with 14,102 more rows</span></pre></div>
<p>If you want to know what genes have been used by CONGAS and to which segments these have been mapped you can use function</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="co"># All mapped genes</span>
<span class="fu"><a href="../reference/get_mapped_genes.html">get_mapped_genes</a></span><span class="op">(</span><span class="va">rcongas_example</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 7,044 x 4</span>
<span class="co">#&gt;    gene     chr   from     to       </span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt;    &lt;chr&gt;    </span>
<span class="co">#&gt;  1 DHRS3    chr1  12600001 156000000</span>
<span class="co">#&gt;  2 HNRNPCL1 chr1  12600001 156000000</span>
<span class="co">#&gt;  3 PRDM2    chr1  12600001 156000000</span>
<span class="co">#&gt;  4 KAZN     chr1  12600001 156000000</span>
<span class="co">#&gt;  5 TMEM51   chr1  12600001 156000000</span>
<span class="co">#&gt;  6 EFHD2    chr1  12600001 156000000</span>
<span class="co">#&gt;  7 DNAJC16  chr1  12600001 156000000</span>
<span class="co">#&gt;  8 AGMAT    chr1  12600001 156000000</span>
<span class="co">#&gt;  9 DDI2     chr1  12600001 156000000</span>
<span class="co">#&gt; 10 PLEKHM2  chr1  12600001 156000000</span>
<span class="co">#&gt; # … with 7,034 more rows</span>

<span class="co"># Genes mapped in chromosomes '1' and '5'</span>
<span class="fu"><a href="../reference/get_mapped_genes.html">get_mapped_genes</a></span><span class="op">(</span><span class="va">rcongas_example</span>, chromosomes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'chr1'</span>, <span class="st">'chr5'</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 1,205 x 4</span>
<span class="co">#&gt;    gene     chr   from     to       </span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt;    &lt;chr&gt;    </span>
<span class="co">#&gt;  1 DHRS3    chr1  12600001 156000000</span>
<span class="co">#&gt;  2 HNRNPCL1 chr1  12600001 156000000</span>
<span class="co">#&gt;  3 PRDM2    chr1  12600001 156000000</span>
<span class="co">#&gt;  4 KAZN     chr1  12600001 156000000</span>
<span class="co">#&gt;  5 TMEM51   chr1  12600001 156000000</span>
<span class="co">#&gt;  6 EFHD2    chr1  12600001 156000000</span>
<span class="co">#&gt;  7 DNAJC16  chr1  12600001 156000000</span>
<span class="co">#&gt;  8 AGMAT    chr1  12600001 156000000</span>
<span class="co">#&gt;  9 DDI2     chr1  12600001 156000000</span>
<span class="co">#&gt; 10 PLEKHM2  chr1  12600001 156000000</span>
<span class="co">#&gt; # … with 1,195 more rows</span></pre></div>
</div>
<div id="post-clustering-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#post-clustering-analysis" class="anchor"></a>Post-clustering analysis</h2>
<ul>
<li>
<code>get_DE_table</code> to obtain results from DE analysis;</li>
</ul>
<p>Results from Differential Expression (DE) analysis using <code>Seurat</code> are available by using <code>get_DE_table</code>. By default there are returned sorted by adjusted p-value, and filters can be applied to subset by maximum adjusted p-value (<code>cut_pvalue</code>), absolute log-fold change (<code>cut_lfc</code>), and chromosome id.</p>
<p>Note that the filters are applied jointly (i.e., as intersection).</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="co"># Default results (sorted by adjusted p-value, default filters)</span>
<span class="fu"><a href="../reference/get_DE_table.html">get_DE_table</a></span><span class="op">(</span><span class="va">rcongas_example</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 132 x 8</span>
<span class="co">#&gt;    chr        from       to gene    pct_detec_1 ptc_detec_2 p_val_adj avg_log2FC</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;    &lt;int&gt; &lt;chr&gt;         &lt;dbl&gt;       &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 chrX   71475529   7.15e7 RPS4X         1           1      9.12e-50     -0.483</span>
<span class="co">#&gt;  2 chrX    1505045   1.51e6 SLC25A6       0.997       1      6.03e-47     -0.688</span>
<span class="co">#&gt;  3 chrX   73040486   7.31e7 XIST          0.068       0.748  3.25e-26     -1.79 </span>
<span class="co">#&gt;  4 chr15  66790355   6.68e7 RPL4          1           1      4.23e-26      0.482</span>
<span class="co">#&gt;  5 chr16   2012053   2.01e6 RPS2          1           1      7.45e-23      0.615</span>
<span class="co">#&gt;  6 chr14 102547075   1.03e8 HSP90A…       1           1      3.46e-21      0.765</span>
<span class="co">#&gt;  7 chr18  43664110   4.37e7 ATP5A1        0.995       0.933  9.63e-21      0.785</span>
<span class="co">#&gt;  8 chr6   29909037   2.99e7 HLA-A         1           1      8.27e-19     -0.793</span>
<span class="co">#&gt;  9 chr15  78832747   7.88e7 PSMA4         0.997       0.941  4.18e-18      0.834</span>
<span class="co">#&gt; 10 chr8  124510129   1.25e8 FBXO32        0.501       0.822  1.57e-17     -1.36 </span>
<span class="co">#&gt; # … with 122 more rows</span>

<span class="co"># DE filtered by p-value cut at alpha-level 5%</span>
<span class="fu"><a href="../reference/get_DE_table.html">get_DE_table</a></span><span class="op">(</span><span class="va">rcongas_example</span>, cut_pvalue <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 222 x 8</span>
<span class="co">#&gt;    chr        from       to gene    pct_detec_1 ptc_detec_2 p_val_adj avg_log2FC</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;    &lt;int&gt; &lt;chr&gt;         &lt;dbl&gt;       &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 chrX   71475529   7.15e7 RPS4X         1           1      9.12e-50     -0.483</span>
<span class="co">#&gt;  2 chrX    1505045   1.51e6 SLC25A6       0.997       1      6.03e-47     -0.688</span>
<span class="co">#&gt;  3 chrX   73040486   7.31e7 XIST          0.068       0.748  3.25e-26     -1.79 </span>
<span class="co">#&gt;  4 chr15  66790355   6.68e7 RPL4          1           1      4.23e-26      0.482</span>
<span class="co">#&gt;  5 chr16   2012053   2.01e6 RPS2          1           1      7.45e-23      0.615</span>
<span class="co">#&gt;  6 chr14 102547075   1.03e8 HSP90A…       1           1      3.46e-21      0.765</span>
<span class="co">#&gt;  7 chr18  43664110   4.37e7 ATP5A1        0.995       0.933  9.63e-21      0.785</span>
<span class="co">#&gt;  8 chr6   29909037   2.99e7 HLA-A         1           1      8.27e-19     -0.793</span>
<span class="co">#&gt;  9 chr15  78832747   7.88e7 PSMA4         0.997       0.941  4.18e-18      0.834</span>
<span class="co">#&gt; 10 chr8  124510129   1.25e8 FBXO32        0.501       0.822  1.57e-17     -1.36 </span>
<span class="co">#&gt; # … with 212 more rows</span>

<span class="co"># DE filtered to retain all values, regardless the LFC</span>
<span class="fu"><a href="../reference/get_DE_table.html">get_DE_table</a></span><span class="op">(</span><span class="va">rcongas_example</span>, cut_lfc <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 135 x 8</span>
<span class="co">#&gt;    chr        from       to gene    pct_detec_1 ptc_detec_2 p_val_adj avg_log2FC</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;    &lt;int&gt; &lt;chr&gt;         &lt;dbl&gt;       &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 chrX   71475529   7.15e7 RPS4X         1           1      9.12e-50     -0.483</span>
<span class="co">#&gt;  2 chrX    1505045   1.51e6 SLC25A6       0.997       1      6.03e-47     -0.688</span>
<span class="co">#&gt;  3 chrX   73040486   7.31e7 XIST          0.068       0.748  3.25e-26     -1.79 </span>
<span class="co">#&gt;  4 chr15  66790355   6.68e7 RPL4          1           1      4.23e-26      0.482</span>
<span class="co">#&gt;  5 chr16   2012053   2.01e6 RPS2          1           1      7.45e-23      0.615</span>
<span class="co">#&gt;  6 chr14 102547075   1.03e8 HSP90A…       1           1      3.46e-21      0.765</span>
<span class="co">#&gt;  7 chr18  43664110   4.37e7 ATP5A1        0.995       0.933  9.63e-21      0.785</span>
<span class="co">#&gt;  8 chr6   29909037   2.99e7 HLA-A         1           1      8.27e-19     -0.793</span>
<span class="co">#&gt;  9 chr15  78832747   7.88e7 PSMA4         0.997       0.941  4.18e-18      0.834</span>
<span class="co">#&gt; 10 chr8  124510129   1.25e8 FBXO32        0.501       0.822  1.57e-17     -1.36 </span>
<span class="co">#&gt; # … with 125 more rows</span>

<span class="co"># A view on a certain chromosome </span>
<span class="fu"><a href="../reference/get_DE_table.html">get_DE_table</a></span><span class="op">(</span><span class="va">rcongas_example</span>, chromosomes <span class="op">=</span> <span class="st">'chr15'</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 15 x 8</span>
<span class="co">#&gt;    chr        from        to gene   pct_detec_1 ptc_detec_2 p_val_adj avg_log2FC</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt; &lt;chr&gt;        &lt;dbl&gt;       &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 chr15  66790355  66816870 RPL4         1           1      4.23e-26      0.482</span>
<span class="co">#&gt;  2 chr15  78832747  78841604 PSMA4        0.997       0.941  4.18e-18      0.834</span>
<span class="co">#&gt;  3 chr15  79102829  79190475 MORF4…       0.984       0.956  1.21e-16      0.708</span>
<span class="co">#&gt;  4 chr15  45003675  45011075 B2M          1           1      3.07e-14     -0.607</span>
<span class="co">#&gt;  5 chr15 101821715 101835487 SNRPA1       0.94        0.8    1.54e-12      0.754</span>
<span class="co">#&gt;  6 chr15  65822756  65870687 PTPLA…       0.973       0.904  6.12e-11      0.681</span>
<span class="co">#&gt;  7 chr15  85212775  85259947 SEC11A       0.97        0.793  1.79e-10      0.617</span>
<span class="co">#&gt;  8 chr15  81239667  81282219 MESDC2       0.699       0.356  6.90e- 8      0.624</span>
<span class="co">#&gt;  9 chr15  78570177  78592136 WDR61        0.737       0.467  4.51e- 7      0.596</span>
<span class="co">#&gt; 10 chr15  64979772  64995480 OAZ2         0.892       0.726  1.87e- 6      0.558</span>
<span class="co">#&gt; 11 chr15  75212132  75230509 COX5A        0.837       0.622  3.59e- 6      0.557</span>
<span class="co">#&gt; 12 chr15  78632666  78640572 CRABP1       0.764       0.496  1.59e- 5      0.648</span>
<span class="co">#&gt; 13 chr15  90626277  90645736 IDH2         0.745       0.481  2.55e- 5      0.578</span>
<span class="co">#&gt; 14 chr15  89010684  89022222 MRPS11       0.756       0.556  4.59e- 5      0.549</span>
<span class="co">#&gt; 15 chr15  75628232  75634268 COMMD4       0.713       0.444  8.44e- 4      0.512</span></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://sites.google.com/view/cdslab">Salvatore Milite</a>, <a href="https://sites.google.com/view/cdslab">Giulio Caravagna</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
