<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Monosomy 7 analysis • Rcongas</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Monosomy 7 analysis">
<meta property="og:description" content="Rcongas">
<meta property="og:image" content="militeee.github.io/Rcongas/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Rcongas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Getters.html">Getters</a>
    </li>
    <li>
      <a href="../articles/Plotting_functions.html">Plotting functions</a>
    </li>
    <li>
      <a href="../articles/gbm_example.html">Glioblastoma smart-seq analysis</a>
    </li>
    <li>
      <a href="../articles/monosomy_example.html">Monosomy 7 analysis</a>
    </li>
    <li>
      <a href="../articles/simulated_example.html">CONGAS on simulated data</a>
    </li>
    <li>
      <a href="../articles/xeno_examples.html">analysis on 10x data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/militeee/Rcongas/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="monosomy_example_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Monosomy 7 analysis</h1>
                        <h4 class="author">Salvatore Milite</h4>
            
            <h4 class="date">2021-01-30</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/militeee/Rcongasvignettes/monosomy_example.Rmd"><code>vignettes/monosomy_example.Rmd</code></a></small>
      <div class="hidden name"><code>monosomy_example.Rmd</code></div>

    </div>

    
    
<p>Here we analyze data from <span class="citation">Zhao et al. (2017)</span>. Data is publicly available at [<a href="https://www.ebi.ac.uk/gxa/sc/experiments/E-CURD-6/downloads" class="uri">https://www.ebi.ac.uk/gxa/sc/experiments/E-CURD-6/downloads</a>], we just changed the metadata file name to ‘metadata.tsv’ and the data matrix names to be readable for Seurat. Than the rest of the analysis closely follow the Seurat vignette on pbmc dataset ([<a href="https://satijalab.org/seurat/v3.2/pbmc3k_tutorial.html" class="uri">https://satijalab.org/seurat/v3.2/pbmc3k_tutorial.html</a>])</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Not really sure how the data were analyzed, I think it's worth to repeat it and have the data preproccessed as you want </span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span> <span class="co"># general scRNA-seq functions</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">biomaRt</span><span class="op">)</span> <span class="co"># annotations and stuff</span>
<span class="co"># Donwnload the files from the EBI database</span>
<span class="co"># You can call the metadata file as you like</span>
<span class="co"># But remember the actual mtx 10x files shoud have this structure</span>
<span class="co"># directory -&gt; matrix.mtx (with the actual counts, Sparse matrix format), barcodes.tsv (with cellular barcodes) and genes.tsv (with gene names)</span>
<span class="va">meta</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"metadata.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="co"># You give the name of the directory where you have the mtx, barcodes and genes file</span>
<span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/Read10X.html">Read10X</a></span><span class="op">(</span>data.dir <span class="op">=</span> <span class="st">"raw/"</span><span class="op">)</span>
<span class="co"># Convert the identifiers from ENSEMBL to HGNC gene names</span>
<span class="va">ensembl</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span><span class="st">"ensembl"</span>,dataset<span class="op">=</span><span class="st">"hsapiens_gene_ensembl"</span><span class="op">)</span>
<span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>attributes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'ensembl_gene_id'</span>,<span class="st">'hgnc_symbol'</span><span class="op">)</span>, 
      filters <span class="op">=</span> <span class="st">'ensembl_gene_id'</span>, 
      values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, 
      mart <span class="op">=</span> <span class="va">ensembl</span><span class="op">)</span>
<span class="va">res</span> <span class="op">=</span> <span class="va">res</span><span class="op">[</span><span class="va">res</span><span class="op">$</span><span class="va">hgnc_symbol</span> <span class="op">!=</span> <span class="st">""</span>,,<span class="op">]</span>
<span class="va">data</span> <span class="op">&lt;-</span>  <span class="va">data</span><span class="op">[</span><span class="va">res</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">=</span> <span class="va">res</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>
<span class="co"># Create a seurat object</span>
<span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">data</span>, min.cells <span class="op">=</span> <span class="fl">3</span>, min.features <span class="op">=</span> <span class="fl">200</span> <span class="op">)</span></code></pre></div>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scdata</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/PercentageFeatureSet.html">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">scdata</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">scdata</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="monosomy_example_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Very high quality dataset, with 10x you go with mt.fract &gt; 15/20 and nFeatures &gt; 200-300</span>
<span class="co">#scdata &lt;- subset(scdata, subset = nFeature_RNA &gt; 7500 &amp; nFeature_RNA &lt; 15000 &amp; percent.mt &lt; 15)</span>
<span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/subset.Seurat.html">subset</a></span><span class="op">(</span><span class="va">scdata</span>, subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">7500</span> <span class="op">&amp;</span> <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">15000</span> <span class="op">&amp;</span> <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">15</span> <span class="op">&amp;</span> <span class="va">nCount_RNA</span> <span class="op">&gt;</span> <span class="fl">6.8e+6</span> <span class="op">&amp;</span> <span class="va">nCount_RNA</span> <span class="op">&lt;</span> <span class="fl">2.4e+7</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">scdata</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="monosomy_example_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span>
<span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span><span class="va">scdata</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">200</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/VariableFeaturePlot.html">VariableFeaturePlot</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Transformation introduced infinite values in continuous x-axis</code></pre>
<p><img src="monosomy_example_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">scdata</span><span class="op">)</span></code></pre></div>
<pre><code>## Centering and scaling data matrix</code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">scdata</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/VariableFeatures.html">VariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">scdata</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## PC_ 1 
## Positive:  CD79B, AKAP12, DNTT, EBF1, RAG1, PTPRE, TP53INP1, ARPP21, CSGALNACT1, DTX1 
##     MYLK, RUBCNL, LCN6, WASHC4, CD9, SH2D4B, E2F2, BACH2, SOX4, RIC1 
##     IRF4, SMAD1, EIF2AK3, LCN10, CHST15, VSIG10, VLDLR, JCHAIN, IL16, EVL 
## Negative:  CA1, HBB, HBD, CA2, HBA1, HBA2, H4C3, IARS1, MPL, SAT1 
##     CXCL8, ACLY, THBS1, MARS2, CLP1, HBG2, DHX58, TESK2, FIG4, C16orf70 
##     PLEK, ZNF30, LYZ, TAF5, ZNF607, HJURP, ADGRG6, ZNF155, ZBTB34, ZNF318 
## PC_ 2 
## Positive:  CA1, HBB, HBA1, CA2, REXO2, HBD, HBA2, AKAP12, TP53INP1, CD79B 
##     RAG1, H4C3, DNTT, E2F2, SMAD1, RIC1, EBF1, ARPP21, YOD1, TESK2 
##     SH2D4B, DTX1, IARS1, USPL1, MARCHF7, LCN6, MT-CO1, HJURP, EIF2AK3, LCN10 
## Negative:  CCL5, NKG7, CCL4, GNLY, GZMB, CD3G, GZMH, TNFAIP3, ITGAX, TRAC 
##     PLEK, MPL, PRDM1, RGS1, DUSP5, FAM107B, LYZ, CAMK2D, IRF8, GRAP2 
##     FBXO32, S100A10, AXL, CD3D, PIK3R5, SPIRE1, RUNX2, GCH1, GPR183, JAML 
## PC_ 3 
## Positive:  CCL4, NKG7, CCL5, GNLY, HBB, GZMB, HBA1, CA1, GZMH, HBA2 
##     CD3G, CA2, HBD, TRAC, LBH, REXO2, GRAP2, FBXO32, TNFAIP3, ITGAX 
##     ISG20, E2F2, CD3D, AKAP12, CSGALNACT1, CEMIP2, CD79B, RAG1, GPCPD1, ZNF18 
## Negative:  MPL, SOX4, PHACTR1, CXCL8, THBS1, LYZ, PLEK, PPP1R16B, RUNX2, AXL 
##     ZNF211, ADGRG6, SNED1, RGS1, ZBTB34, S100A10, CAPG, IRF8, JAML, CYBB 
##     YIPF3, ZNF823, OTULINL, NAPSB, TPP1, SAT1, GPR183, FGF7, TMEM176B, SRP68 
## PC_ 4 
## Positive:  MPL, THBS1, PHACTR1, CXCL8, GABPB1, CCL5, ZNF211, MARCHF7, CEMIP2, SNED1 
##     VSIG10, YIPF3, STRIP1, SPIRE1, AKAP12, TP53INP1, PIK3IP1, BRAP, TSC1, CD9 
##     CD3G, ZNF57, ADGRG6, ZNF266, SMAD1, IARS1, C16orf70, RAG1, ZNF33A, FGF7 
## Negative:  IRF8, JCHAIN, RUNX2, GPR183, CYBB, S100A10, IGKC, AXL, LYZ, NAPSB 
##     OTULINL, CAPG, PTPRE, DUSP5, H4C3, FAM107B, BEND5, JAML, PIK3R5, RUBCNL 
##     SULF2, CRYBG1, RGS1, HBB, NCKIPSD, HBA1, CA1, MPO, HBA2, HBD 
## PC_ 5 
## Positive:  LYZ, AXL, S100A10, GPR183, TPP1, PLEK, NAPSB, SULF2, SAT1, CXCL8 
##     OTULINL, THBS1, ITGAX, JAK1, REXO2, PHACTR1, MPL, MRFAP1, SMAD1, WASHC4 
##     KAT5, CA2, TBC1D25, HBA1, AKAP12, HBA2, RAG1, CD9, CAMK2D, CA1 
## Negative:  JCHAIN, DNTT, CAPG, CYBB, EBF1, TNFAIP3, RGS1, FIG4, IL16, ADGRG6 
##     KBTBD4, CDH2, CHST15, C11orf95, SCMH1, SUCNR1, IGKC, RUNX2, DICER1, ABCA12 
##     PRR4, TMEM168, NAXD, SOX4, ZNF550, H4C3, MARS2, IRF8, TRAT1, TAF5</code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">scdata</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></code></pre></div>
<p><img src="monosomy_example_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">scdata</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></code></pre></div>
<pre><code>## Computing nearest neighbor graph</code></pre>
<pre><code>## Computing SNN</code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">scdata</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 598
## Number of edges: 37115
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.6495
## Number of communities: 3
## Elapsed time: 0 seconds</code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">scdata</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
## To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
## This message will be shown once per session</code></pre>
<pre><code>## 22:38:56 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
<pre><code>## 22:38:56 Read 598 rows and found 30 numeric columns</code></pre>
<pre><code>## 22:38:56 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
<pre><code>## 22:38:56 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
<pre><code>## 0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
<pre><code>## [----|----|----|----|----|----|----|----|----|----|</code></pre>
<pre><code>## **************************************************|
## 22:38:57 Writing NN index file to temp file /tmp/RtmpghLMlh/file654666733abd
## 22:38:57 Searching Annoy index using 1 thread, search_k = 3000
## 22:38:57 Annoy recall = 100%
## 22:38:57 Commencing smooth kNN distance calibration using 1 thread
## 22:38:58 Initializing from normalized Laplacian + noise
## 22:38:58 Commencing optimization for 500 epochs, with 22330 positive edges
## 22:39:00 Optimization finished</code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">new_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">meta</span><span class="op">$</span><span class="va">V4</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">new_meta</span><span class="op">)</span> <span class="op">&lt;-</span>  <span class="va">meta</span><span class="op">$</span><span class="va">V1</span>
<span class="va">scdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/AddMetaData.html">AddMetaData</a></span><span class="op">(</span><span class="va">scdata</span>,<span class="va">new_meta</span>, <span class="st">"patient"</span><span class="op">)</span>
<span class="co">#create a new object with specific patients, for successive analysis,  you can also subset based on the UMAP clusters</span>
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">scdata</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"patient"</span><span class="op">)</span></code></pre></div>
<p><img src="monosomy_example_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scdata_filt</span> <span class="op">&lt;-</span> <span class="va">scdata</span><span class="op">[</span>,<span class="va">scdata</span><span class="op">$</span><span class="va">patient</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"patient 1"</span>, <span class="st">"H4"</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">scdata</span>, features <span class="op">=</span> <span class="st">"PLCB4"</span><span class="op">)</span></code></pre></div>
<p><img src="monosomy_example_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>As we do not have any prior segmentation nor SNPs VAF we can download a cytobad file from UCSC and try to run CONGAS over entire chromosomes (remember that we are looking for big aneuploidy events).</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># I load rcongas from source, you can just install the package from github with </span>
<span class="co"># devtools::install_github("militeee/Rcongas")</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Rcongas</span><span class="op">)</span></code></pre></div>
<pre><code>## Registered S3 method overwritten by 'cli':
##   method     from    
##   print.boxx spatstat</code></pre>
<pre><code>## ✓ Loading Rcongas, 'Copy-Number genotyping from single cells'. Support : &lt;https://militeee.github.io/Rcongas/&gt;</code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></code></pre></div>
<pre><code>## 
## Attaching package: 'dplyr'</code></pre>
<pre><code>## The following object is masked from 'package:biomaRt':
## 
##     select</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union</code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Cromosomal arms information</span>
<span class="co">#curl::curl_download("https://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/cytoBand.txt.gz", destfile = "./cytoband.tsv.gz")</span>
<span class="co"># Hope you are on a UNIX system</span>
<span class="co">#system("gzip -d cytoband.tsv.gz")</span>
<span class="va">cyto</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"cytoband.tsv"</span><span class="op">)</span>

<span class="co"># CONGAS wants a coloumn for prior expectation on the CN value </span>
<span class="va">cyto</span> <span class="op">&lt;-</span>  <span class="va">cyto</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">V1</span>, <span class="va">V2</span>, <span class="va">V3</span>, <span class="va">V4</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>V4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html">substr</a></span><span class="op">(</span><span class="va">V4</span>, <span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">V1</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">V1</span><span class="op">)</span>, start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">V2</span><span class="op">)</span>, end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">V3</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">chr</span>, <span class="va">start</span>, <span class="va">end</span><span class="op">)</span><span class="op">%&gt;%</span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>tot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">chr</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We can retrieve the count matrix from the Seurat object</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#extract the different data you want from the seurat object</span>
<span class="va">data_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">scdata_filt</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">data</span><span class="op">)</span>
<span class="va">data_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">scdata_filt</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>With dataset having a high number of heterogeneous cellular types, median filtering genes over a chromosome with a short window can be useful to remove effects of cell type specific markers with a lot of count that could potentially create false positive. To do so we define a custom function to be passed to get_data that does this filtering and then sums the counts.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">filter_median_by_segs</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">k</span> <span class="op">=</span> <span class="fl">3</span>, <span class="va">na.rm</span> <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">{</span>
  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>
  <span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/runmed.html">runmed</a></span><span class="op">(</span><span class="va">x</span>,k <span class="op">=</span> <span class="va">k</span>, endrule <span class="op">=</span> <span class="st">"constant"</span><span class="op">)</span>
  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">x</span>,na.rm <span class="op">=</span> <span class="va">na.rm</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span> </code></pre></div>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">express_counts</span> <span class="op">&lt;-</span> <span class="fu">Rcongas</span><span class="fu">:::</span><span class="fu">get_data</span><span class="op">(</span><span class="va">data_mat</span>, cnv_data <span class="op">=</span> <span class="va">cyto</span>, type <span class="op">=</span> <span class="st">"fixed_binning"</span>, fun <span class="op">=</span> <span class="va">filter_median_by_segs</span>, correct_bins <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="co"># always take a look at the heatmap </span>
<span class="va">express_counts</span> <span class="op">&lt;-</span> <span class="fu">Rcongas</span><span class="fu">:::</span><span class="fu">filter_segments.rcongas</span><span class="op">(</span>X <span class="op">=</span> <span class="va">express_counts</span>, filter_mu <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></code></pre></div>
<pre><code>##  chr1:0:249250621  chr2:0:243199373  chr3:0:198022430  chr4:0:191154276 
##              TRUE              TRUE              TRUE              TRUE 
##  chr5:0:180915260  chr6:0:171115067  chr7:0:159138663  chr8:0:146364022 
##              TRUE              TRUE              TRUE              TRUE 
##  chr9:0:141213431 chr10:0:135534747 chr11:0:135006516 chr12:0:133851895 
##              TRUE              TRUE              TRUE              TRUE 
## chr13:0:115169878 chr14:0:107349540 chr15:0:102531392  chr16:0:90354753 
##              TRUE              TRUE              TRUE              TRUE 
##  chr17:0:81195210  chr18:0:78077248  chr19:0:59128983  chr20:0:63025520 
##              TRUE              TRUE              TRUE              TRUE 
##  chr21:0:48129895  chr22:0:51304566 
##             FALSE              TRUE</code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">express_counts</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">/</span><span class="va">express_counts</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">cnv</span><span class="op">$</span><span class="va">mu</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="va">t</span><span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">express_counts</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span>, cluster_cols <span class="op">=</span> <span class="cn">F</span>,  show_rownames <span class="op">=</span> <span class="cn">F</span>, annotation_row <span class="op">=</span> <span class="va">new_meta</span>, color <span class="op">=</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fl">9</span>,<span class="st">"Reds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="monosomy_example_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># if you feel like scaling, you may also used the scaled genes from the Seurat object as shown above, never tried but more sense than this to me </span>
<span class="co">#express$data$counts &lt;- scale(express$data$counts, T, T)</span></code></pre></div>
<p>We are now ready to perform the inference. Note how we are using different theta hyperparameters here compared to the 10x example.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">express_counts</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">express_counts</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>

<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">Rcongas</span><span class="fu">:::</span><span class="fu">best_cluster</span><span class="op">(</span><span class="va">express_counts</span>,model <span class="op">=</span> <span class="st">"MixtureGaussian"</span>, clusters <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, steps <span class="op">=</span> <span class="fl">300</span>, lr <span class="op">=</span> <span class="fl">0.01</span>, param_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"theta_scale"</span> <span class="op">=</span> <span class="fl">3.5</span>, <span class="st">"theta_rate"</span> <span class="op">=</span><span class="fl">0.01</span>, <span class="st">"cnv_var"</span> <span class="op">=</span> <span class="fl">0.10</span><span class="op">)</span> , method <span class="op">=</span> <span class="st">"BIC"</span>,  MAP <span class="op">=</span> <span class="cn">T</span>, posteriors <span class="op">=</span> <span class="cn">T</span>, step_post <span class="op">=</span> <span class="fl">150</span><span class="op">)</span>
<span class="co"># filter small clusters</span>
<span class="va">res_filt</span> <span class="op">&lt;-</span>  <span class="fu">Rcongas</span><span class="fu">:::</span><span class="fu">filter_clusters</span><span class="op">(</span><span class="va">res</span>, ncells <span class="op">=</span> <span class="fl">10</span>, remove <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#plotting the results with some annotations</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span>

<span class="va">bc</span> <span class="op">&lt;-</span>  <span class="va">res_filt</span><span class="op">$</span><span class="va">inference</span><span class="op">$</span><span class="va">model_selection</span><span class="op">$</span><span class="va">best_K</span>

<span class="va">new_meta2</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">new_meta</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">res_filt</span><span class="op">$</span><span class="va">inference</span><span class="op">$</span><span class="va">models</span><span class="op">[[</span><span class="va">bc</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">parameters</span><span class="op">$</span><span class="va">assignement</span><span class="op">)</span>,, drop <span class="op">=</span> <span class="cn">F</span><span class="op">]</span>

<span class="va">new_meta2</span><span class="op">$</span><span class="st">"Log total counts"</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">scdata_filt</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">res_filt</span><span class="op">$</span><span class="va">inference</span><span class="op">$</span><span class="va">models</span><span class="op">[[</span><span class="va">bc</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">parameters</span><span class="op">$</span><span class="va">assignement</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">nCount_RNA</span><span class="op">)</span>

<span class="va">new_meta2</span><span class="op">$</span><span class="va">CONGAS</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">res_filt</span><span class="op">$</span><span class="va">inference</span><span class="op">$</span><span class="va">models</span><span class="op">[[</span><span class="va">bc</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">parameters</span><span class="op">$</span><span class="va">assignement</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">new_meta2</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Sample ID"</span>

<span class="va">p1</span> <span class="op">=</span> <span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">express_counts</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">res_filt</span><span class="op">$</span><span class="va">inference</span><span class="op">$</span><span class="va">models</span><span class="op">[[</span><span class="va">bc</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">parameters</span><span class="op">$</span><span class="va">assignement</span><span class="op">)</span>,<span class="op">]</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">/</span><span class="va">express_counts</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">cnv</span><span class="op">$</span><span class="va">mu</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="va">t</span> ,<span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">/</span> <span class="va">res_filt</span><span class="op">$</span><span class="va">inference</span><span class="op">$</span><span class="va">models</span><span class="op">[[</span><span class="va">bc</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">parameters</span><span class="op">$</span><span class="va">norm_factor</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="va">res_filt</span><span class="op">$</span><span class="va">inference</span><span class="op">$</span><span class="va">models</span><span class="op">[[</span><span class="va">bc</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">parameters</span><span class="op">$</span><span class="va">assignement</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span>, cluster_cols <span class="op">=</span> <span class="cn">F</span>,  show_rownames <span class="op">=</span> <span class="cn">F</span>, annotation_row <span class="op">=</span> <span class="va">new_meta2</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span><span class="fl">11</span>, <span class="st">"RdBu"</span><span class="op">)</span><span class="op">)</span>, cluster_rows <span class="op">=</span> <span class="cn">F</span>, gaps_row <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">res_filt</span><span class="op">$</span><span class="va">inference</span><span class="op">$</span><span class="va">models</span><span class="op">[[</span><span class="va">bc</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">parameters</span><span class="op">$</span><span class="va">assignement</span><span class="op">)</span><span class="op">)</span> <span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span><span class="op">]</span>,breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">3</span>, length.out <span class="op">=</span> <span class="fl">11</span><span class="op">)</span>
<span class="op">)</span>
<span class="va">p1</span></code></pre></div>
<p><img src="monosomy_example_files/figure-html/unnamed-chunk-14-1.png" width="700"> Than we can again perform DE and plot it.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_DE.html">calculate_DE</a></span><span class="op">(</span><span class="va">res_filt</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">data_mat</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">data_mat</span><span class="op">)</span> <span class="op">%in%</span> <span class="op">(</span><span class="fu">Rcongas</span><span class="fu">::</span><span class="va">hg38_gene_coordinates</span> <span class="op">%&gt;%</span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">chr</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chrX"</span>, <span class="st">"chrY"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">res_filt</span><span class="op">$</span><span class="va">inference</span><span class="op">$</span><span class="va">models</span><span class="op">[[</span><span class="va">bc</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">parameters</span><span class="op">$</span><span class="va">assignement</span><span class="op">)</span><span class="op">]</span>, clone1 <span class="op">=</span> <span class="st">"c1"</span>, <span class="st">"c2"</span>, method <span class="op">=</span> <span class="st">"DESeq2"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_DE_volcano.html">plot_DE_volcano</a></span><span class="op">(</span><span class="va">de</span><span class="op">)</span></code></pre></div>
<pre><code>## ✓ Loading CNAqc, 'Copy Number Alteration quality check'. Support : &lt;https://caravagn.github.io/CNAqc/&gt;</code></pre>
<p><img src="monosomy_example_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-1k" class="csl-entry">
Zhao, Xin, Shouguo Gao, Zhijie Wu, Sachiko Kajigaya, Xingmin Feng, Qingguo Liu, Danielle M Townsley, et al. 2017. <span>“Single-Cell RNA-Seq Reveals a Distinct Transcriptome Signature of Aneuploid Hematopoietic Cells.”</span> <em>Blood</em> 130 (25): 2762–73.
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://sites.google.com/view/cdslab">Salvatore Milite</a>, <a href="https://sites.google.com/view/cdslab">Giulio Caravagna</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
