<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>analysis on 10x data • Rcongas</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="analysis on 10x data">
<meta property="og:description" content="Rcongas">
<meta property="og:image" content="militeee.github.io/Rcongas/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Rcongas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Getters.html">Getters</a>
    </li>
    <li>
      <a href="../articles/Plotting_functions.html">Plotting functions</a>
    </li>
    <li>
      <a href="../articles/gbm_example.html">Glioblastoma smart-seq analysis</a>
    </li>
    <li>
      <a href="../articles/monosomy_example.html">Monosomy 7 analysis</a>
    </li>
    <li>
      <a href="../articles/simulated_example.html">CONGAS on simulated data</a>
    </li>
    <li>
      <a href="../articles/xeno_examples.html">analysis on 10x data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/militeee/Rcongas/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="xeno_examples_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>analysis on 10x data</h1>
            
            <h4 class="date">2021-01-30</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/militeee/Rcongasvignettes/xeno_examples.Rmd"><code>vignettes/xeno_examples.Rmd</code></a></small>
      <div class="hidden name"><code>xeno_examples.Rmd</code></div>

    </div>

    
    
<p>This is an example analysis with a dataset of trple-negative breast cancer, sequenced with 10x. The data has been obtained from Campbell at al 2019 and is currently available at [<a href="https://zenodo.org/record/2363826#.YBQHtsIo9hE" class="uri">https://zenodo.org/record/2363826#.YBQHtsIo9hE</a>]</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Read teh data</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://Matrix.R-forge.R-project.org/">Matrix</a></span><span class="op">)</span>

<span class="va">counts_SA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/externalFormats.html">readMM</a></span><span class="op">(</span><span class="st">"clonealign-processed-data/SA501/10X/20171026_SA501X2XB00096/outs/filtered_gene_bc_matrices/hg19/matrix.mtx"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"clonealign-processed-data/SA501/10X/20171026_SA501X2XB00096/outs/filtered_gene_bc_matrices/hg19/genes.tsv"</span><span class="op">)</span>
  
<span class="va">barcodes</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"clonealign-processed-data/SA501/10X/20171026_SA501X2XB00096/outs/filtered_gene_bc_matrices/hg19/barcodes.tsv"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">counts_SA</span><span class="op">)</span> <span class="op">&lt;-</span>  <span class="va">genes</span><span class="op">$</span><span class="va">V2</span>

<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">counts_SA</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">barcodes</span><span class="op">$</span><span class="va">V1</span></code></pre></div>
<pre><code>## Loading Rcongas</code></pre>
<pre><code>## ✓ Loading Rcongas, 'Copy-Number genotyping from single cells'. Support : &lt;https://militeee.github.io/Rcongas/&gt;</code></pre>
<p>We first do a bit of preprocessing as filtering for cells with few expressed genes</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"rcongas"</span><span class="op">)</span>

<span class="va">mat_pre</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="fu">preprocess_sc</span><span class="op">(</span>mat <span class="op">=</span> <span class="va">counts_SA</span>, filter_upper_quantile <span class="op">=</span> <span class="cn">T</span>, perc_cells_gene_expr <span class="op">=</span> <span class="fl">0.05</span>,upper_quantile <span class="op">=</span> <span class="fl">0.95</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Then we load the CN profile of the most abundat clone, and we consider it was a result obtained from bulkDNA analysis (as this clone A compose around the 80% of the total cells, we suppose an hypothetical CNV inference from bulkDNA to mostly represent its profile).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#read the clones data and add chr on front of the names</span>

<span class="va">clones_dataA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"clonealign-processed-data/SA501/cnv/SA501X3F-cluster_A.segments.csv"</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">7</span><span class="op">)</span><span class="op">]</span>
<span class="va">clones_dataA</span><span class="op">$</span><span class="va">tot</span> <span class="op">&lt;-</span>  <span class="va">clones_dataA</span><span class="op">$</span><span class="va">integer_copy_number</span>
<span class="va">clones_dataA</span><span class="op">$</span><span class="va">chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="va">clones_dataA</span><span class="op">$</span><span class="va">chr</span><span class="op">)</span>

<span class="co"># Aggregate counts on segments</span>
<span class="va">inp2</span> <span class="op">&lt;-</span>  <span class="fu">get_data</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">mat_pre</span><span class="op">)</span>, cnv_data <span class="op">=</span> <span class="va">clones_dataA</span>, type <span class="op">=</span> <span class="st">"fixed_binning"</span>, startsWithchr <span class="op">=</span> <span class="cn">F</span>,  correct_bins <span class="op">=</span> <span class="cn">T</span>, genome <span class="op">=</span> <span class="st">"hg19"</span>, custom_gene_matrix <span class="op">=</span> <span class="va">counts_SA</span><span class="op">)</span>
<span class="co">#Filter for lowly expressed segments </span>
<span class="va">inp3</span> <span class="op">&lt;-</span> <span class="fu">filter_segments.rcongas</span><span class="op">(</span>X <span class="op">=</span> <span class="va">inp2</span>, filter_mu <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># exploratory analysis </span>
<span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="op">(</span><span class="op">(</span><span class="va">inp3</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">inp3</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span> <span class="op">)</span> <span class="op">/</span> <span class="va">inp3</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">cnv</span><span class="op">$</span><span class="va">ploidy_real</span>, show_rownames <span class="op">=</span> <span class="cn">F</span>, show_colnames <span class="op">=</span> <span class="cn">T</span>, cluster_cols <span class="op">=</span> <span class="cn">F</span>, cutree_rows <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<p><img src="xeno_examples_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>After an exploratory analysis with an heatmap we can run CONGAS. We first run it to calculate the MAP values for the CNV LogNormal distributions. We then filter for small clusters.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>
<span class="co"># Perform the inference</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">best_cluster</span><span class="op">(</span><span class="va">inp3</span>,model <span class="op">=</span> <span class="st">"MixtureGaussian"</span>, clusters <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, steps <span class="op">=</span> <span class="fl">800</span>, lr <span class="op">=</span> <span class="fl">0.01</span>, param_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"theta_scale"</span> <span class="op">=</span> <span class="fl">3.2</span>, <span class="st">"theta_rate"</span> <span class="op">=</span> <span class="fl">1</span>, <span class="st">"cnv_var"</span> <span class="op">=</span> <span class="fl">0.50</span><span class="op">)</span> , method <span class="op">=</span> <span class="st">"BIC"</span>,  MAP <span class="op">=</span> <span class="cn">T</span>, posteriors <span class="op">=</span> <span class="cn">T</span>, step_post <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>
<span class="co">#Filter for small clusters</span>
<span class="va">res_filt</span> <span class="op">&lt;-</span>  <span class="fu">filter_clusters</span><span class="op">(</span><span class="va">res</span>, abundance <span class="op">=</span> <span class="fl">0.03</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#counts plot</span>
<span class="fu"><a href="../reference/plot_counts_rna_segments.html">plot_counts_rna_segments</a></span><span class="op">(</span><span class="va">res_filt</span>, z_score <span class="op">=</span> <span class="cn">T</span>, normalised <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<pre><code>## ✓ Loading CNAqc, 'Copy Number Alteration quality check'. Support : &lt;https://caravagn.github.io/CNAqc/&gt;</code></pre>
<p><img src="xeno_examples_files/figure-html/unnamed-chunk-7-1.png" width="700"> After that we can use our MAP estimates as a starting point to learn the full mean and variance of the LogNormals.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># rerun with full Bayesian model</span>
<span class="va">rerun</span> <span class="op">&lt;-</span>  <span class="fu">run_complete</span><span class="op">(</span><span class="va">res_filt</span><span class="op">)</span></code></pre></div>
<p>RCONGAS comes with a function to quickly calculate differentially expressed genes between two clones.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># differential expression</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="va">de</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_DE.html">calculate_DE</a></span><span class="op">(</span><span class="va">rerun</span>, <span class="va">counts_SA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">mat_pre</span><span class="op">)</span>,<span class="op">]</span> <span class="op">%&gt;%</span>  <span class="va">t</span>, clone1 <span class="op">=</span> <span class="st">"1"</span>, clone2 <span class="op">=</span> <span class="st">"2"</span>, method <span class="op">=</span> <span class="st">"DESeq2"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_DE_volcano.html">plot_DE_volcano</a></span><span class="op">(</span><span class="va">de</span><span class="op">)</span></code></pre></div>
<p><img src="xeno_examples_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://sites.google.com/view/cdslab">Salvatore Milite</a>, <a href="https://sites.google.com/view/cdslab">Giulio Caravagna</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
