<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Create a dataset. — init • Rcongas</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Create a dataset. — init" />
<meta property="og:description" content="This function creates a dataset (an object of class rcongasplus) by assembling multiple single-cell input measurements
(ATAC and/or RNA data modalities), the input segmentation (from bulk DNA sequencing),
and the per-cell normalisation factors for the data.
All input data are passed as tibbles; the input formats are as follows:
for single-cell ATAC/RNA data, the cell identifier, the genomic coordinates
(chr, from, to) which refer either to an ATAC peak, or an RNA gene
identifier, and a value reporting the reads mapped.
for the input segmentation, the genomic coordinates
(chr, from, to) which refer to the segment, and the number of
copies (i.e., DNA ploidy) of the segment.
for normalization factors the cell identifier, the actual normalisation_factor
and the modality to wihch the factor refers to


This function receives also other parameters - e.g., the models likelihoods - which
will determine the overall behaviour of the underlying model, and how data are preared for inference.
A Negative Binomial likelihood (&quot;NB&quot;), which works directly from raw counts data
A Gaussian likelihood (&quot;G&quot;), which requires a z-score transformation of the data. This consists
in :
scaling raw counts by the input normalization factors;
computing z-scores per cell;
summing up z-scores per segment;
computing z-scores per segment;
center the z-scores mean to the input ploidy.


" />
<meta property="og:image" content="caravagnalab.github.io/Rcongas/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Rcongas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Rcongas.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Filtering.html">Filtering functions</a>
    </li>
    <li>
      <a href="../articles/Plotting.html">Plotting functions</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/caravagnalab/Rcongas/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create a dataset.</h1>
    <small class="dont-index">Source: <a href='https://github.com/caravagnalab/RcongasR/init.R'><code>R/init.R</code></a></small>
    <div class="hidden name"><code>init.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function creates a dataset (an object of class <code>rcongasplus</code>) by assembling multiple single-cell input measurements
(ATAC and/or RNA data modalities), the input segmentation (from bulk DNA sequencing),
and the per-cell normalisation factors for the data.</p>
<p>All input data are passed as tibbles; the input formats are as follows:</p><ul>
<li><p>for single-cell ATAC/RNA data, the <code>cell</code> identifier, the genomic coordinates
(<code>chr</code>, <code>from</code>, <code>to</code>) which refer either to an ATAC peak, or an RNA gene
identifier, and a <code>value</code> reporting the reads mapped.</p></li>
<li><p>for the input segmentation, the genomic coordinates
(<code>chr</code>, <code>from</code>, <code>to</code>) which refer to the segment, and the number of
<code>copies</code> (i.e., DNA ploidy) of the segment.</p></li>
<li><p>for normalization factors the <code>cell</code> identifier, the actual <code>normalisation_factor</code>
and the <code>modality</code> to wihch the factor refers to</p></li>
</ul>

<p>This function receives also other parameters - e.g., the models likelihoods - which
will determine the overall behaviour of the underlying model, and how data are preared for inference.</p><ul>
<li><p>A Negative Binomial likelihood (<code>"NB"</code>), which works directly from raw counts data</p></li>
<li><p>A Gaussian likelihood (<code>"G"</code>), which requires a z-score transformation of the data. This consists
in :</p><ul>
<li><p>scaling raw counts by the input normalization factors;</p></li>
<li><p>computing z-scores per cell;</p></li>
<li><p>summing up z-scores per segment;</p></li>
<li><p>computing z-scores per segment;</p></li>
<li><p>center the z-scores mean to the input ploidy.</p></li>
</ul></li>
</ul>

    </div>

    <pre class="usage"><span class='fu'>init</span><span class='op'>(</span>
  <span class='va'>rna</span>,
  <span class='va'>atac</span>,
  <span class='va'>segmentation</span>,
  rna_normalisation_factors <span class='op'>=</span> <span class='va'>rna</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='auto_normalisation_factor.html'>auto_normalisation_factor</a></span><span class='op'>(</span><span class='op'>)</span>,
  atac_normalisation_factors <span class='op'>=</span> <span class='va'>atac</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='auto_normalisation_factor.html'>auto_normalisation_factor</a></span><span class='op'>(</span><span class='op'>)</span>,
  rna_likelihood <span class='op'>=</span> <span class='st'>"G"</span>,
  atac_likelihood <span class='op'>=</span> <span class='st'>"NB"</span>,
  reference_genome <span class='op'>=</span> <span class='st'>"GRCh38"</span>,
  description <span class='op'>=</span> <span class='st'>"(R)CONGAS+ model"</span>,
  smooth <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>rna</th>
      <td><p>A tibble with single-cell RNA data.</p></td>
    </tr>
    <tr>
      <th>atac</th>
      <td><p>A tibble with single-cell ATAC data.</p></td>
    </tr>
    <tr>
      <th>segmentation</th>
      <td><p>A tibble with the input segmentation.</p></td>
    </tr>
    <tr>
      <th>rna_normalisation_factors</th>
      <td><p>The RNA tibble with the input per-cell normalisation factors.
By default these are computed by function <code>auto_normalisation_factor</code>.</p></td>
    </tr>
    <tr>
      <th>atac_normalisation_factors</th>
      <td><p>The ATAC tibble with the input per-cell normalisation factors.
By default these are computed by function <code>auto_normalisation_factor</code>.</p></td>
    </tr>
    <tr>
      <th>rna_likelihood</th>
      <td><p>Type of likelihood used for RNA data (<code>"G"</code> for Gaussian and
<code>""NB</code> for Negative Binomial). The RNA default is <code>"G"</code>.</p></td>
    </tr>
    <tr>
      <th>atac_likelihood</th>
      <td><p>Type of likelihood used for ATAC data, with default <code>"NB"</code>.</p></td>
    </tr>
    <tr>
      <th>reference_genome</th>
      <td><p>Either <code>"GRCh38"</code> or <code>"hg19"</code>.</p></td>
    </tr>
    <tr>
      <th>description</th>
      <td><p>A model in-words description.</p></td>
    </tr>
    <tr>
      <th>smooth</th>
      <td><p>If yes, input segments are smootheed by joining per chromosome segments that
have the same ploidy.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An object of class <code>rcongasplus</code></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"example_input"</span><span class='op'>)</span>

<span class='co'># For instance, RNA data</span>
<span class='va'>example_input</span><span class='op'>$</span><span class='va'>x_rna</span> <span class='op'>%&gt;%</span> <span class='va'>print</span>
</div><div class='output co'>#&gt; NULL</div><div class='input'>
<span class='co'># .. or ATAC data</span>
<span class='va'>example_input</span><span class='op'>$</span><span class='va'>x_atac</span> <span class='op'>%&gt;%</span> <span class='va'>print</span>
</div><div class='output co'>#&gt; NULL</div><div class='input'>
<span class='co'># .. and segmentation</span>
<span class='va'>example_input</span><span class='op'>$</span><span class='va'>x_segmentation</span> <span class='op'>%&gt;%</span> <span class='va'>print</span>
</div><div class='output co'>#&gt; NULL</div><div class='input'>
<span class='co'># .. and normalisation factors can be computed (default)</span>
<span class='va'>example_input</span><span class='op'>$</span><span class='va'>x_rna</span> <span class='op'>%&gt;%</span> <span class='fu'><a href='auto_normalisation_factor.html'>auto_normalisation_factor</a></span><span class='op'>(</span><span class='op'>)</span>
</div><div class='output co'>#&gt; NULL</div><div class='input'>
<span class='va'>x</span> <span class='op'>=</span> <span class='fu'>init</span><span class='op'>(</span>
  rna <span class='op'>=</span> <span class='va'>example_input</span><span class='op'>$</span><span class='va'>x_rna</span>,
  atac <span class='op'>=</span> <span class='va'>example_input</span><span class='op'>$</span><span class='va'>x_atac</span>,
  segmentation <span class='op'>=</span> <span class='va'>example_input</span><span class='op'>$</span><span class='va'>x_segmentation</span>,
  rna_likelihood <span class='op'>=</span> <span class='st'>"G"</span>,
  atac_likelihood <span class='op'>=</span> <span class='st'>'NB'</span>,
  description <span class='op'>=</span> <span class='st'>'My model'</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='error'>Error in init(rna = example_input$x_rna, atac = example_input$x_atac,     segmentation = example_input$x_segmentation, rna_likelihood = "G",     atac_likelihood = "NB", description = "My model"): Cannot have both assays null.</span></div><div class='input'>
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>── <span style='color: #000000;background-color: #BBBB00;'>[ (R)CONGAS+ ]</span><span style='color: #0000BB;'> Zaho et al. MDS hematopoietical used in CONGAS (HCA)</span><span> ─────────</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>── CNA segments (reference: </span><span style='color: #00BB00;'>GRCh38</span><span>) </span></div><div class='output co'>#&gt; <span class='message'>→ Input </span><span style='color: #00BB00;'>21</span><span> CNA segments, mean ploidy </span><span style='color: #00BB00;'>2</span><span>.</span></div><div class='output co'>#&gt; 
#&gt; 	</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>|</span><span style='background-color: #00BB00;'> </span><span>
#&gt; 
#&gt; 	 </span><span style='text-decoration: underline;'>Ploidy:</span><span>  </span><span style='background-color: #00BBBB;'> </span><span> 0   </span><span style='background-color: #0000BB;'> </span><span> 1   </span><span style='background-color: #00BB00;'> </span><span> 2   </span><span style='background-color: #BBBB00;'> </span><span> 3   </span><span style='background-color: #BB0000;'> </span><span> 4   </span><span style='background-color: #BB00BB;'> </span><span> 5   </span><span style='background-color: #BBBBBB;'> </span><span> *   </div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>── Modalities </span></div><div class='output co'>#&gt; <span class='message'>→ RNA: </span><span style='color: #00BB00;'>100</span><span> cells with </span><span style='color: #00BB00;'>18841</span><span> mapped genes, </span><span style='color: #00BB00;'>890478</span><span> non-zero values. Likelihood: </span><span style='color: #00BB00;'>Gaussian</span><span>.</span></div><div class='output co'>#&gt; <span class='message'>→ ATAC: </span><span style='color: #BB0000;'>not available</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #BBBB00;'>!</span><span> Clusters: </span><span style='color: #BB0000;'>not available</span><span>.</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>── </span><span style='color: #BBBBBB;background-color: #BB00BB;font-weight: bold;'> LOG </span><span> ──</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; - 2021-04-02 19:34:18 Created input object.</div><div class='output co'>#&gt; [1] 0</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://www.caravagnalab.org'>Salvatore Milite</a>, <a href='https://www.caravagnalab.org'>Riccardo Bergamin</a>, <a href='https://www.caravagnalab.org'>Giulio Caravagna</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


